{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From /beagle3/roux/kyleghaby/kininhibs/main/\n",
    "\n",
    "16 Dec 2025"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "50188fb0d5f545cdafedac31c856ca41",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import re\n",
    "#import mdtraj\n",
    "from shutil import copyfile\n",
    "import os\n",
    "import sys\n",
    "from IPython.display import Markdown as md\n",
    "import math\n",
    "import numpy as np\n",
    "import parmed as pmd #https://github.com/ParmEd/ParmEd\n",
    "import nglview #https://github.com/nglviewer/nglview\n",
    "#vmd (linux ver) also required\n",
    "#AMBERTOOLS (cpptraj) required\n",
    "#pdb_tools also required\n",
    "#updated toppar: http://mackerell.umaryland.edu/charmm_ff.shtml#charmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/c/Users/Kyle/Desktop/Roux/kininhibs/main/TAK1_SM1/build'"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.mkdir('build')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('build')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#notes in README\n",
    "#filled residues: "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 1. Download and process PDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "- keep xtal waters with the protein\n",
    "- run the code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-02-16 11:09:42--  https://www.rcsb.org/fasta/entry/5J7S\n",
      "Resolving www.rcsb.org (www.rcsb.org)... 128.6.159.248\n",
      "Connecting to www.rcsb.org (www.rcsb.org)|128.6.159.248|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 465 [text/x-fasta]\n",
      "Saving to: ‘5J7S.fasta’\n",
      "\n",
      "5J7S.fasta          100%[===================>]     465  --.-KB/s    in 0s      \n",
      "\n",
      "2023-02-16 11:09:42 (43.5 MB/s) - ‘5J7S.fasta’ saved [465/465]\n",
      "\n",
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) http://www.ks.uiuc.edu/Research/vmd/                         \n",
      "Info) Email questions and bug reports to vmd@ks.uiuc.edu           \n",
      "Info) Please include this reference in published work using VMD:   \n",
      "Info)    Humphrey, W., Dalke, A. and Schulten, K., `VMD - Visual   \n",
      "Info)    Molecular Dynamics', J. Molec. Graphics 1996, 14.1, 33-38.\n",
      "Info) -------------------------------------------------------------\n",
      "Info) Multithreading available, 12 CPUs.\n",
      "Info)   CPU features: SSE2 SSE4.1 AVX AVX2 FMA F16 HT \n",
      "Info) Free system memory: 10GB (62%)\n",
      "Info) No CUDA accelerator devices available.\n",
      "Info) Dynamically loaded 3 plugins in directory:\n",
      "Info) /usr/local/lib/vmd/plugins/LINUXAMD64/molfile\n",
      "after#0\n",
      "after#1\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl /mnt/c/Users/Kyle/Desktop/utilities/vmdStore\n",
      "The vmdStore package could not be loaded:\n",
      "\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl /mnt/c/Users/Kyle/Desktop/utilities/vmdStore /mnt/c/Users/Kyle/Desktop/utilities/vmdStore/plugins/molUP\n",
      "The molUP package could not be loaded:\n",
      "\n",
      "Info) Using plugin pdb for structure file 5J7S.pdb\n",
      "Info) Using plugin pdb for coordinates from file 5J7S.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 4467\n",
      "Info)    Bonds: 4535\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 280\n",
      "Info)    Waters: 0\n",
      "Warning) Unusual bond between residues:  174 (protein) and 601 (none)\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 5   Protein: 5   Nucleic: 0\n",
      "Info) Finished with coordinate file 5J7S.pdb.\n",
      "0\n",
      "Info) Opened coordinate file 5J7S_noaltloc.pdb for writing.\n",
      "Info) Finished with coordinate file 5J7S_noaltloc.pdb.\n",
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) Exiting normally.\n",
      "vmd > "
     ]
    }
   ],
   "source": [
    "PDBID='5J7S'\n",
    "ligResname='6H3'\n",
    "\n",
    "#fetch pdb and fasta, renumber residues to start with 1, and discard alternative coordinates\n",
    "!./../../code/get_xtal.sh {PDBID}\n",
    "\n",
    "!mkdir -p modeller\n",
    "!mkdir -p scwrl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 2. Polish protein structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "- Fill in missing residues in xtal/{PDBID}_noaltloc.pdb using Modeller in Chimera\n",
    "    1. Open PDB file, then show its Sequence (Chimera menu: Tools… Sequence… Sequence) or skip to step D if you have the full sequence file already\n",
    "    2. Save the sequence as FASTA file (sequence window menu: File… Save As, choose file type: aligned FASTA)\n",
    "    3. Edit the FASTA file to add the new sequence\n",
    "    4. Quit from the first sequence window and open the new fasta file in Chimera.\n",
    "    5. Associate it with the structure (sequence window menu: Structure… Associations, change from \"none\" to the new sequence)\n",
    "    6. Use the Modeller interface (sequence window menu:  Structure… Modeller (loops/refinement)) to model \"non-terminal missing structure\" or \"all missing structure\". Make sure this also fills in missing side chains of all residues.            \n",
    "        -  Note for terminal regions of kinases: \n",
    "           Cterm does not affect kinase activity so leave as is. N term might be important. The tryptophan (AKDAW) is needed bc wedge (W260 in Src (2OIQ)) and D258 is probably also important bc salt bridge. Most include AKDAW. Good to aim for including AKDAW.\n",
    "    7. Save output as modeller/{PDBID}_noaltloc.pdb.\n",
    "    8. While modeller is open, note the side chains that were incomplete. SCWRL is used to remodel the side chains for those residues.\n",
    "\n",
    "- Renumber residues and make a new pdb with just prot/wat (code below)\n",
    "    - not strictly necessary since modeller output will likely only have protein and water and will be renumbered, but it helps to rename the files so later steps can expect those files to exist.\n",
    "- Model side chains not in xtal with SCWRL (code below)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "#renumber residues to start with 1\n",
    "!pdb_reres modeller/{PDBID}_noaltloc.pdb > modeller/{PDBID}_noaltloc_reres.pdb \n",
    "\n",
    "#remove lig (and other residues) from prot/wat (assuming pdb only contains prot,lig,wat)\n",
    "!pdb_delresname -{ligResname} modeller/{PDBID}_noaltloc_reres.pdb  > modeller/prot_wat.pdb\n",
    "\n",
    "#make a ligand-only pdb file in xtal\n",
    "!pdb_selresname -{ligResname} xtal/{PDBID}_noaltloc.pdb  > xtal/{ligResname}_noaltloc.pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====XTAL(267)==== cvkieevigagevcrgrlkqpgrrevfvaiktlkvgyterqrrdflseasimgqfdhpniirlegvvtksrpvmiltefmencaldsflrlndgqftviqlvgmlrgiaagmkylsemnyvhrdlaarnilvnsnlvckvsdfgleddpsdptytsslggkipirwtapeaiayrkftsasdvwsygivmwevmsygerpywdmsnqdvinaveqdyrlpppmdcptalhqlmldcwvrdrnlrpkfsqivntldklirnpaslkvi \n",
      "\n",
      "====MODELLER(274)==== cvkieevigagefgevcrgrlkqpgrrevfvaiktlkvgyterqrrdflseasimgqfdhpniirlegvvtksrpvmiltefmencaldsflrlndgqftviqlvgmlrgiaagmkylsemnyvhrdlaarnilvnsnlvckvsdfglsrfleddpsdptytsslggkipirwtapeaiayrkftsasdvwsygivmwevmsygerpywdmsnqdvinaveqdyrlpppmdcptalhqlmldcwvrdrnlrpkfsqivntldklirnpaslkvi \n",
      "\n",
      "====GAPS(274)==== cvkieevigaGEFgevcrgrlkqpgrrevfvaiktlkvgyterqrrdflseasimgqfdhpniirlegvvtksrpvmiltefmencaldsflrlndgqftviqlvgmlrgiaagmkylsemnyvhrdlaarnilvnsnlvckvsdfglSRFLeddpsdptytsslggkipirwtapeaiayrkftsasdvwsygivmwevmsygerpywdmsnqdvinaveqdyrlpppmdcptalhqlmldcwvrdrnlrpkfsqivntldklirnpaslkvi \n",
      "\n",
      "Residues with incomplete side chains: 636 (5), 637 (6), 651 (20), 657 (26), 658 (27), 659 (28), 668 (37), 669 (38), 672 (41), 673 (42), 674 (43), 677 (46), 703 (72), 705 (74) \n",
      "\n",
      "====INCOMPLETE SIDE CHAINS(267)==== cvkiEEvigagevcrgrlkQpgrreVFVaiktlkvgYTerQRRdfLseasimgqfdhpniirlegvvtksrPvMiltefmencaldsflrlndgqftviqlvgmlrgiaagmkylsemnyvhrdlaarnilvnsnlvckvsdfgleddpsdptytsslggkipirwtapeaiayrkftsasdvwsygivmwevmsygerpywdmsnqdvinaveqdyrlpppmdcptalhqlmldcwvrdrnlrpkfsqivntldklirnpaslkvi \n",
      "\n",
      "====SCWRL(274)==== cvkiEEvigaGEFgevcrgrlkQpgrreVFVaiktlkvgYTerQRRdfLseasimgqfdhpniirlegvvtksrPvMiltefmencaldsflrlndgqftviqlvgmlrgiaagmkylsemnyvhrdlaarnilvnsnlvckvsdfglSRFLeddpsdptytsslggkipirwtapeaiayrkftsasdvwsygivmwevmsygerpywdmsnqdvinaveqdyrlpppmdcptalhqlmldcwvrdrnlrpkfsqivntldklirnpaslkvi \n",
      "\n",
      "Check ambiguous residues and edit scwrl/scwrl.seq to change capitalization if needed. Ambigious residues are those with adjacent repeats, leading to ambiguity about the exact location of the gap. This ambiguity can be resolved by checking with residue is linked in the PDB.\n"
     ]
    }
   ],
   "source": [
    "# Generate SCWRL input\n",
    "\n",
    "def find_incomplete_residues(pdb_file, amino_acids):\n",
    "    '''\n",
    "    Function to detect residues with incomplete side chains in PDB\n",
    "    '''\n",
    "    incomplete_residues = set()\n",
    "    current_residue = None\n",
    "    current_atoms = set()\n",
    "    first_res=None\n",
    "    with open(pdb_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('ATOM'):\n",
    "                res_name = line[17:20].strip()\n",
    "                res_seq = int(line[22:26].strip())\n",
    "                atom_name = line[12:16].strip()\n",
    "                if not first_res:\n",
    "                    first_res=res_seq\n",
    "                if current_residue != (res_name, res_seq):\n",
    "                    if current_residue:\n",
    "                        expected_atoms = set(amino_acids.get(current_residue[0], []))\n",
    "                        if not expected_atoms.issubset(current_atoms):\n",
    "                            incomplete_residues.add(current_residue[1])\n",
    "\n",
    "                    current_residue = (res_name, res_seq)\n",
    "                    current_atoms = set()\n",
    "\n",
    "                current_atoms.add(atom_name)\n",
    "\n",
    "    # Check last residue\n",
    "    if current_residue:\n",
    "        expected_atoms = set(amino_acids.get(current_residue[0], []))\n",
    "        if not expected_atoms.issubset(current_atoms):\n",
    "            incomplete_residues.add(current_residue[1])\n",
    " \n",
    "    sorted_set=sorted(incomplete_residues)\n",
    "    incomplete_residues={num - (first_res - 1) for num in sorted_set}\n",
    "    print('Residues with incomplete side chains:',', '.join(f\"{num} ({num - (first_res - 1)})\" for num in sorted_set),'\\n')\n",
    "    return incomplete_residues\n",
    "\n",
    "\n",
    "def align_and_capitalize(sequence_with_gaps, sequence_with_incomplete):\n",
    "    \"\"\"\n",
    "    Aligns two sequences and returns the first sequence with residues capitalized based on the second sequence.\n",
    "    \"\"\"\n",
    "    # Initialize pointers and output list\n",
    "    ptr_gaps = 0\n",
    "    ptr_incomplete = 0\n",
    "    output = []\n",
    "    \n",
    "    # Loop through the sequence with gaps\n",
    "    while ptr_gaps < len(sequence_with_gaps):\n",
    "        residue_gaps = sequence_with_gaps[ptr_gaps]\n",
    "        \n",
    "        # Handle unmodeled residues (capitalized in sequence_with_gaps)\n",
    "        if residue_gaps.isupper():\n",
    "            output.append(residue_gaps)\n",
    "            ptr_gaps += 1\n",
    "            continue\n",
    "        \n",
    "        # Loop through the sequence with incomplete side chains\n",
    "        while ptr_incomplete < len(sequence_with_incomplete):\n",
    "            residue_incomplete = sequence_with_incomplete[ptr_incomplete]\n",
    "            \n",
    "            # Match found\n",
    "            if residue_gaps.lower() == residue_incomplete.lower():\n",
    "                # Capitalize if the residue in sequence_with_incomplete is capitalized\n",
    "                if residue_incomplete.isupper():\n",
    "                    output.append(residue_gaps.upper())\n",
    "                else:\n",
    "                    output.append(residue_gaps.lower())\n",
    "                \n",
    "                ptr_incomplete += 1\n",
    "                break\n",
    "            \n",
    "            ptr_incomplete += 1\n",
    "        \n",
    "        ptr_gaps += 1\n",
    "    \n",
    "    return ''.join(output)\n",
    "\n",
    "def extract_seq_and_residues(pdb_lines):\n",
    "    '''\n",
    "    Function to extract sequence and residue numbers from a PDB file\n",
    "    '''\n",
    "    seq = []\n",
    "    residues = []\n",
    "    for line in pdb_lines:\n",
    "        if line.startswith(\"ATOM\") and line[13:15].strip() == \"CA\":\n",
    "            res_num = int(line[22:26].strip())\n",
    "            amino_acid = line[17:20].strip()\n",
    "            # Using standard amino acid codes (IUPAC)\n",
    "            aa_dict = {'ALA': 'A', 'CYS': 'C', 'ASP': 'D', 'GLU': 'E', 'PHE': 'F',\n",
    "                       'GLY': 'G', 'HIS': 'H', 'ILE': 'I', 'LYS': 'K', 'LEU': 'L',\n",
    "                       'MET': 'M', 'ASN': 'N', 'PRO': 'P', 'GLN': 'Q', 'ARG': 'R',\n",
    "                       'SER': 'S', 'THR': 'T', 'VAL': 'V', 'TRP': 'W', 'TYR': 'Y'}\n",
    "            seq.append(aa_dict.get(amino_acid, 'X'))\n",
    "            residues.append(res_num)\n",
    "    return ''.join(seq), residues\n",
    "\n",
    "def find_extensions_and_insertions(xtal_seq, xtal_residues, modeller_seq):\n",
    "    # Create a dictionary to map xtal residues to their positions \n",
    "    xtal_map = {res: aa for res, aa in zip(xtal_residues, xtal_seq)}\n",
    "    \n",
    "    # Generate gap-filled sequence for xtal_seq\n",
    "    gap_filled_xtal_seq = []\n",
    "    for i in range(min(xtal_residues), max(xtal_residues) + 1):\n",
    "        gap_filled_xtal_seq.append(xtal_map.get(i, '-'))\n",
    "\n",
    "    # Shift modeller_seq and identify insertions\n",
    "    max_overlap = 0\n",
    "    best_shift = 0\n",
    "    for shift in range(len(modeller_seq) - len(gap_filled_xtal_seq) + 1):\n",
    "        overlap = 0\n",
    "        for i in range(len(gap_filled_xtal_seq)):\n",
    "            if gap_filled_xtal_seq[i] == modeller_seq[i + shift] or gap_filled_xtal_seq[i] == '-':\n",
    "                overlap += 1\n",
    "        if overlap > max_overlap:\n",
    "            max_overlap = overlap\n",
    "            best_shift = shift\n",
    "\n",
    "    # Generate the output sequence \n",
    "    output_seq = []\n",
    "    for i in range(len(modeller_seq)):\n",
    "        if i < best_shift or i >= best_shift + len(gap_filled_xtal_seq):\n",
    "            output_seq.append(modeller_seq[i].upper())\n",
    "        else:\n",
    "            if gap_filled_xtal_seq[i - best_shift] == '-':\n",
    "                output_seq.append(modeller_seq[i].upper())\n",
    "            else:\n",
    "                output_seq.append(modeller_seq[i])\n",
    "\n",
    "    return ''.join(output_seq)\n",
    "\n",
    "# Define amino acid side-chain atoms \n",
    "amino_acids = {\n",
    "    'ALA': ['CB'],\n",
    "    'ARG': ['CB', 'CG', 'CD', 'NE', 'CZ', 'NH1', 'NH2'],\n",
    "    'ASN': ['CB', 'CG', 'OD1', 'ND2'],\n",
    "    'ASP': ['CB', 'CG', 'OD1', 'OD2'],\n",
    "    'CYS': ['CB', 'SG'],\n",
    "    'GLN': ['CB', 'CG', 'CD', 'OE1', 'NE2'],\n",
    "    'GLU': ['CB', 'CG', 'CD', 'OE1', 'OE2'],\n",
    "    'GLY': [],\n",
    "    'HIS': ['CB', 'CG', 'ND1', 'CD2', 'CE1', 'NE2'],\n",
    "    'ILE': ['CB', 'CG1', 'CG2', 'CD1'],\n",
    "    'LEU': ['CB', 'CG', 'CD1', 'CD2'],\n",
    "    'LYS': ['CB', 'CG', 'CD', 'CE', 'NZ'],\n",
    "    'MET': ['CB', 'CG', 'SD', 'CE'],\n",
    "    'PHE': ['CB', 'CG', 'CD1', 'CD2', 'CE1', 'CE2', 'CZ'],\n",
    "    'PRO': ['CB', 'CG', 'CD'],\n",
    "    'SER': ['CB', 'OG'],\n",
    "    'THR': ['CB', 'OG1', 'CG2'],\n",
    "    'TRP': ['CB', 'CG', 'CD1', 'CD2', 'NE1', 'CE2', 'CE3', 'CZ2', 'CZ3', 'CH2'],\n",
    "    'TYR': ['CB', 'CG', 'CD1', 'CD2', 'CE1', 'CE2', 'CZ', 'OH'],\n",
    "    'VAL': ['CB', 'CG1', 'CG2']\n",
    "}\n",
    "\n",
    "\n",
    "# Load PDB and Fasta files\n",
    "xtal_pdb = f'xtal/{PDBID}_noaltloc.pdb'\n",
    "modeller_pdb = f'modeller/prot_wat.pdb'\n",
    "\n",
    "# Extract sequences and residue numbers\n",
    "xtal_pdb_lines = [line.strip() for line in open(xtal_pdb, 'r')]\n",
    "modeller_pdb_lines = [line.strip() for line in open(modeller_pdb, 'r')]\n",
    "xtal_seq, xtal_residues = extract_seq_and_residues(xtal_pdb_lines)\n",
    "xtal_seq=xtal_seq.lower()\n",
    "print(f\"====XTAL({len(xtal_seq)})====\",xtal_seq,'\\n')\n",
    "modeller_seq, modeller_residues = extract_seq_and_residues(modeller_pdb_lines)\n",
    "modeller_seq=modeller_seq.lower()\n",
    "print(f\"====MODELLER({len(modeller_seq)})====\",modeller_seq,'\\n')\n",
    "\n",
    "# Identify gaps and extensions\n",
    "sequence_with_gaps = find_extensions_and_insertions(xtal_seq, xtal_residues, modeller_seq)\n",
    "print(f\"====GAPS({len(sequence_with_gaps)})====\", sequence_with_gaps,'\\n')\n",
    "\n",
    "# Identify incomplete residues\n",
    "incomplete_residues = find_incomplete_residues(xtal_pdb, amino_acids)\n",
    "sequence_with_incomplete = \"\"\n",
    "for i, aa in enumerate(xtal_seq):\n",
    "    if i + 1 in incomplete_residues:\n",
    "        sequence_with_incomplete += aa.upper()\n",
    "    else:\n",
    "        sequence_with_incomplete += aa.lower()\n",
    "print(f\"====INCOMPLETE SIDE CHAINS({len(sequence_with_incomplete)})====\",sequence_with_incomplete,'\\n')\n",
    "\n",
    "# Generate final SCWRL sequence\n",
    "final_scwrl_sequence = align_and_capitalize(sequence_with_gaps, sequence_with_incomplete)\n",
    "print(f\"====SCWRL({len(final_scwrl_sequence)})====\",final_scwrl_sequence,'\\n')\n",
    "\n",
    "# Write final sequence to scwrl.seq\n",
    "with open('scwrl/scwrl.seq', 'w') as f:\n",
    "    f.write(final_scwrl_sequence)\n",
    "\n",
    "print(\"Check ambiguous residues and edit scwrl/scwrl.seq to change capitalization if needed. Ambigious residues are those with adjacent repeats, leading to ambiguity about the exact location of the gap. This ambiguity can be resolved by checking with residue is linked in the PDB.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Using parameters from /mnt/c/FCCC/Scwrl4_linux//Scwrl4.ini\n",
      "\n",
      "<Timing tag = \"Loading_PDB_file\"  duration = \"0.016\" />\n",
      "\n",
      "Using rotamer library: /mnt/c/FCCC/Scwrl4_linux/bbDepRotLib.bin\n",
      "\n",
      "Residue A   1  is being treated as N-terminal\n",
      "\n",
      "Residue A 306  is being treated as C-terminal\n",
      "\n",
      "<Timing tag = \"Adding_residues\"  duration = \"0.016\" />\n",
      "\n",
      "<Timing tag = \"Loading_FrameFile\"  duration = \"0.000\" />\n",
      "\n",
      "Sequence file has been loaded\n",
      "---------------\n",
      "\"slhmidykeieveevvgrgAFgvvckakwrakdvaikQIESESERkafivelrqlsrvnhpnivklyGAClnpvclvmeyaeggslynvlhgaeplpyytaahamswclqcsqgvaylhsmqpkalihrdlkppnlllvaggtvlkicdfgTACDIQTHMTNNKGsaawmapevfegsnysekcdvfswgiilwevitrrkpfdeiggpafrimwavhngtrppliknlpkpieslmtrcwskdpsqrpsmeeivkimthlmryfpgadeplqypcqhslppgedgrvepyvdfaefyrlwsvdhg\"\n",
      "---------------\n",
      "\n",
      "<Timing tag = \"Building_Atoms_Sets\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Building_Flexible_Subrotamers\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Computing_Symmetry_Operators\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Updating_Self_Energies\"  duration = \"2.922\" />\n",
      "\n",
      "<Timing tag = \"Computing_Active_Rotamers\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Resolving_Disulfide_Bonds\"  duration = \"0.000\" />\n",
      "\n",
      "\n",
      "Total number of subrotamer interactions = 220114\n",
      "\n",
      "<Timing tag = \"Updating_Pairwise_Energies\"  duration = \"0.062\" />\n",
      "\n",
      "<Timing tag = \"Normalizing_Atoms_Sets\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Computing_Interaction_Graph\"  duration = \"2.984\" />\n",
      " \n",
      "\n",
      "<Timing tag = \"Building_Atoms_Sets\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Building_Flexible_Subrotamers\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Computing_Symmetry_Operators\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Updating_Self_Energies\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Computing_Active_Rotamers\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Resolving_Disulfide_Bonds\"  duration = \"0.000\" />\n",
      "\n",
      "\n",
      "Total number of subrotamer interactions = 0\n",
      "\n",
      "<Timing tag = \"Updating_Pairwise_Energies\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Normalizing_Atoms_Sets\"  duration = \"0.000\" />\n",
      "\n",
      "\n",
      "Running Goldstein's DEE (version G)\n",
      ":::::::::::::::::::::::::::::::::::\n",
      "\n",
      "Initial number of residues - 66\n",
      "Initial number of rotamers - 416\n",
      "\n",
      "\t46 redundant edges eliminated\n",
      "\n",
      "\tIsolated residues removed - 50 \n",
      "\n",
      "round 1:\n",
      "\t3 redundant edges eliminated\n",
      "\n",
      "\tIsolated residues removed - 12 \n",
      "\tRotamers eliminated  -  245\n",
      "\n",
      "round 2:\n",
      "\t0 redundant edges eliminated\n",
      "\n",
      "\tIsolated residues removed - 0 \n",
      "\tRotamers eliminated  -  1\n",
      "\n",
      "round 3:\n",
      "\t0 redundant edges eliminated\n",
      "\n",
      "\tIsolated residues removed - 0 \n",
      "\tRotamers eliminated  -  0\n",
      "Energy accumulated during DEE = 9597.32\n",
      "<Timing tag = \"DEE\"  duration = \"0.000\" />\n",
      "\n",
      "Searching for clusters... 2 clusters found\n",
      "\n",
      "Processing cluster of 2 vertices\n",
      "\n",
      "Redundant nodes eliminated : 1\n",
      "Tree-decomposition has been built. Width+1 = 1\n",
      "\n",
      "Total combinations to search = 6\n",
      "   0        :   ->    7  12\n",
      "\n",
      "ArtPts =  0\tMaxDegree =  0\tMaxBag =  1\tAvgBag = 2\n",
      "\n",
      "<Trace tag = \"Deciding_on_the_root_node\" >\n",
      "\n",
      "Number of combinations = 6\n",
      "<Timing tag = \"Deciding_on_the_root_node\"  duration = \"0.000\" />\n",
      "</Trace>\n",
      "<Trace tag = \"Top_To_Bottom\" >\n",
      "\n",
      "<Timing tag = \"Top_To_Bottom\"  duration = \"0.000\" />\n",
      "</Trace>\n",
      "Energy of this cluster = 34.3273\n",
      "\n",
      "\n",
      "Processing cluster of 2 vertices\n",
      "\n",
      "Redundant nodes eliminated : 1\n",
      "Tree-decomposition has been built. Width+1 = 1\n",
      "\n",
      "Total combinations to search = 1e+1\n",
      "   0        :   ->   34  36\n",
      "\n",
      "ArtPts =  0\tMaxDegree =  0\tMaxBag =  1\tAvgBag = 2\n",
      "\n",
      "<Trace tag = \"Deciding_on_the_root_node\" >\n",
      "\n",
      "Number of combinations = 1e+1\n",
      "<Timing tag = \"Deciding_on_the_root_node\"  duration = \"0.000\" />\n",
      "</Trace>\n",
      "<Trace tag = \"Top_To_Bottom\" >\n",
      "\n",
      "<Timing tag = \"Top_To_Bottom\"  duration = \"0.000\" />\n",
      "</Trace>\n",
      "Energy of this cluster = 13.2216\n",
      "\n",
      "\n",
      "Total minimal energy of the graph = 9644.87\n",
      "\n",
      "<Timing tag = \"Combinatorial_Search\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Deriving_Optimum\"  duration = \"0.000\" />\n",
      "\n",
      "<Timing tag = \"Saving_to_PDB_file\"  duration = \"0.016\" />\n",
      "\n",
      "\n",
      "Deleting workspace.\n",
      "\n",
      "<Timing tag = \"Scwrl4_Execution\"  duration = \"3.078\" />\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Run SCWRL command\n",
    "!scwrl4 -i modeller/prot_wat.pdb -s scwrl/scwrl.seq -f xtal/{PDBID}_noaltloc.pdb -o scwrl/prot_wat.pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total waters: 94\n",
      "ATOM: 0, HETATM: 94\n",
      "HOH: 94, WAT: 0\n"
     ]
    }
   ],
   "source": [
    "# Add the xtal waters to scwrl output \n",
    "\n",
    "# Initialize counters\n",
    "atom_count = 0\n",
    "hetatm_count = 0\n",
    "hoh_count = 0\n",
    "wat_count = 0\n",
    "\n",
    "# Open source PDB file for reading and destination PDB file for appending\n",
    "with open('modeller/prot_wat.pdb', 'r') as src, open('scwrl/prot_wat.pdb', 'a') as dest:\n",
    "    for line in src:\n",
    "        # Check for ATOM or HETATM and water residues (HOH or WAT)\n",
    "        if line.startswith('ATOM'):\n",
    "            residue = line[17:20].strip()\n",
    "            if residue in ('HOH', 'WAT'):\n",
    "                atom_count += 1\n",
    "                if residue == 'HOH':\n",
    "                    hoh_count += 1\n",
    "                else:\n",
    "                    wat_count += 1\n",
    "                dest.write(line)\n",
    "        elif line.startswith('HETATM'):\n",
    "            residue = line[17:20].strip()\n",
    "            if residue in ('HOH', 'WAT'):\n",
    "                hetatm_count += 1\n",
    "                if residue == 'HOH':\n",
    "                    hoh_count += 1\n",
    "                else:\n",
    "                    wat_count += 1\n",
    "                dest.write(line)\n",
    "\n",
    "# Logging (counters used for efficient single-pass tallying)\n",
    "print(f\"Total waters: {hoh_count + wat_count}\")\n",
    "print(f\"ATOM: {atom_count}, HETATM: {hetatm_count}\")\n",
    "print(f\"HOH: {hoh_count}, WAT: {wat_count}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Renumber structure\n",
    "def renumber_pdb(file_path):\n",
    "    '''\n",
    "    Function to renumber atoms and residues in a PDB file\n",
    "    '''\n",
    "    atom_counter = 1  # Initialize atom counter\n",
    "    residue_counter = 0  # Initialize residue counter\n",
    "    last_residue = None  # Track last residue ID\n",
    "    \n",
    "    # Read original lines from PDB\n",
    "    with open(file_path, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "    \n",
    "    # Open PDB file for writing\n",
    "    with open(file_path, 'w') as f:\n",
    "        for line in lines:\n",
    "            # Check for ATOM or HETATM lines\n",
    "            if line.startswith(('ATOM', 'HETATM')):\n",
    "                # Update atom number (columns 7-11)\n",
    "                line = line[:6] + f\"{atom_counter:5d}\" + line[11:]\n",
    "                \n",
    "                # Check if the residue ID has changed (columns 23-26)\n",
    "                current_residue = line[22:26].strip()\n",
    "                if current_residue != last_residue:\n",
    "                    residue_counter += 1\n",
    "                last_residue = current_residue\n",
    "                \n",
    "                # Update residue number (columns 23-26)\n",
    "                line = line[:22] + f\"{residue_counter:4d}\" + line[26:]\n",
    "                \n",
    "                # Increment atom counter\n",
    "                atom_counter += 1\n",
    "                \n",
    "            # Write the modified or unmodified line\n",
    "            f.write(line)\n",
    "\n",
    "# File path to the PDB file that needs to be renumbered\n",
    "file_path = 'scwrl/prot_wat.pdb'\n",
    "\n",
    "# Renumber the PDB file\n",
    "renumber_pdb(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) http://www.ks.uiuc.edu/Research/vmd/                         \n",
      "Info) Email questions and bug reports to vmd@ks.uiuc.edu           \n",
      "Info) Please include this reference in published work using VMD:   \n",
      "Info)    Humphrey, W., Dalke, A. and Schulten, K., `VMD - Visual   \n",
      "Info)    Molecular Dynamics', J. Molec. Graphics 1996, 14.1, 33-38.\n",
      "Info) -------------------------------------------------------------\n",
      "Info) Multithreading available, 12 CPUs.\n",
      "Info)   CPU features: SSE2 SSE4.1 AVX AVX2 FMA F16 HT \n",
      "Info) Free system memory: 9GB (55%)\n",
      "Info) No CUDA accelerator devices available.\n",
      "libGL error: No matching fbConfigs or visuals found\n",
      "libGL error: failed to load driver: swrast\n",
      "Info) OpenGL renderer: NVIDIA GeForce GTX 1070 Ti/PCIe/SSE2\n",
      "Info)   Features: STENCIL MSAA(4) MDE MTX NPOT PP PS \n",
      "Info)   GLSL rendering mode is NOT available.\n",
      "Info)   Textures: 2-D (32768x32768), 3-D (16384x16384x16384), Multitexture (4)\n",
      "Info) Dynamically loaded 3 plugins in directory:\n",
      "Info) /usr/local/lib/vmd/plugins/LINUXAMD64/molfile\n",
      "after#0\n",
      "after#1\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/tk8.5 /usr/local/lib/vmd/scripts/tk8.5/ttk /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl /mnt/c/Users/Kyle/Desktop/utilities/vmdStore\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/tk8.5 /usr/local/lib/vmd/scripts/tk8.5/ttk /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl /mnt/c/Users/Kyle/Desktop/utilities/vmdStore /mnt/c/Users/Kyle/Desktop/utilities/vmdStore/plugins/molUP\n",
      "\"xtal/5J7S_noaltloc.pdb\" \"modeller/prot_wat.pdb\" \"scwrl/prot_wat.pdb\"\n",
      " \n",
      "Loading xtal/5J7S_noaltloc.pdb\n",
      "Info) Using plugin pdb for structure file xtal/5J7S_noaltloc.pdb\n",
      "Info) Using plugin pdb for coordinates from file xtal/5J7S_noaltloc.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Finished with coordinate file xtal/5J7S_noaltloc.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 4467\n",
      "Info)    Bonds: 4535\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 280\n",
      "Info)    Waters: 0\n",
      "Warning) Unusual bond between residues:  174 (protein) and 601 (none)\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 5   Protein: 5   Nucleic: 0\n",
      " \n",
      " \n",
      "Loading modeller/prot_wat.pdb\n",
      "Info) Using plugin pdb for structure file modeller/prot_wat.pdb\n",
      "Info) Using plugin pdb for coordinates from file modeller/prot_wat.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Finished with coordinate file modeller/prot_wat.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 2431\n",
      "Info)    Bonds: 2502\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 306\n",
      "Info)    Waters: 0\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 1   Protein: 1   Nucleic: 0\n",
      " \n",
      " \n",
      "Loading scwrl/prot_wat.pdb\n",
      "Info) Using plugin pdb for structure file scwrl/prot_wat.pdb\n",
      "Info) Using plugin pdb for coordinates from file scwrl/prot_wat.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Finished with coordinate file scwrl/prot_wat.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 2954\n",
      "Info)    Bonds: 3040\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 306\n",
      "Info)    Waters: 0\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 1   Protein: 1   Nucleic: 0\n",
      " \n",
      "vmd > Info) In any publication of scientific results based in part or\n",
      "Info) completely on the use of the program STRIDE, please reference:\n",
      "Info)  Frishman,D & Argos,P. (1995) Knowledge-based secondary structure\n",
      "Info)  assignment. Proteins: structure, function and genetics, 23, 566-579.\n",
      "Info) Using plugin pdb for structure file /mnt/c/Users/Kyle/Desktop/Roux/kininhibs/main/TAK1_SM1/build/xtal/5J7S.pdb\n",
      "Info) Using plugin pdb for coordinates from file /mnt/c/Users/Kyle/Desktop/Roux/kininhibs/main/TAK1_SM1/build/xtal/5J7S.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 4467\n",
      "Info)    Bonds: 4535\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 280\n",
      "Info)    Waters: 0\n",
      "Warning) Unusual bond between residues:  174 (protein) and 601 (none)\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 5   Protein: 5   Nucleic: 0\n",
      "Info) Finished with coordinate file /mnt/c/Users/Kyle/Desktop/Roux/kininhibs/main/TAK1_SM1/build/xtal/5J7S.pdb.\n",
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) Exiting normally.\n"
     ]
    }
   ],
   "source": [
    "# check pdbs\n",
    "\n",
    "\n",
    "check='''\n",
    "set filelist {\"xtal/PDBID_noaltloc.pdb\" \"modeller/prot_wat.pdb\" \"scwrl/prot_wat.pdb\"}\n",
    "\n",
    "foreach i $filelist {\n",
    "  echo \" \"\n",
    "  echo \"Loading $i\"\n",
    "  mol new \"$i\" type {pdb} waitfor all\n",
    "  echo \" \"\n",
    "}\n",
    "'''\n",
    "\n",
    "check=check.replace(\"PDBID\",PDBID)\n",
    "\n",
    "with open(\"scwrl/check.tcl\", \"w\") as check_fh:\n",
    "    check_fh.write(check)\n",
    "!vmd -e scwrl/check.tcl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 3. Use CHARMM-GUI solution builder steps 1-3 to make a system comprising water, ions, and protein. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "- Change xtal water segid to WATA so it doesnt clash with SOLV segment used by GUI\n",
    "- Fix resnames\n",
    "    - If aspartic acids are separated from solvent, local pH causes it to be protonated. \n",
    "    - Fix histidine resnames. It may not matter but its correct. HIS to HSD/HSE.\n",
    "        - eg. HIS384 in 2OIQ should be HSE, and the atom name was changed manually. HSE and HSD says which nitrogen is protonated. Programs may default to HSD. Rename to HSD or HSE with the GUI mutation option. but GUI might defualt to HSD anyway so you prob only need to mutate to HSE.\n",
    "- Add disulfides if needed. \n",
    "- Edge distance of 12 fit to protein size (rect)\n",
    "- 0.15 M KCl (intracellular)\n",
    "- Note the PBC used: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dimensions of periodic boundaries established in the precursor steps with charmm-gui. \n",
    "#Format string as comma-delineated string of dimension lengths to be used in chamber box option (eg pbc=\"89,89,89\")\n",
    "pbc=\"93,93,93\"\n",
    "\n",
    "# HIS ___ were mutated into HSE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 4. Align:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "- Old reference pdb with ligand to new step3 pdb (step3 PDB contains coords to fit with PBC).\n",
    "    - pymol\n",
    "- Extract aligned ligand coordinates to replace old ligand coordinates. (code below)\n",
    "    - pdb_tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pdb_selresname -{lig_resname} {PDBID}_aligned2charmm.pdb | pdb_reres > {lig_resname}_aligned2charmm.pdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 5. Make ligand PDBs and mol2's."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "   - Using Pymol:\n",
    "        - Ligand with attached residue with acetylated N-term and N-methylamidated C-term (patch coordinates)\n",
    "            1. Use pymol builder to get linked state, add caps, and change bond orders. Make sure the linking atom has its hydrogens in the right pose relative to the true link so that the linking carbon is tetrahedral. \n",
    "            2. Save as pdb. When saving, enable the following:\n",
    "                - Write multiple bonds as duplicate CONECT records\n",
    "                - Write CONECT records for all bonds\n",
    "                - Write segment identifier (segi) column\n",
    "            3. Put ligand residue first so atomnames in ligand arent replaced by CGenFF. Rename attached residues atomnames with the following nomenclature (unless they clash with ligand atomnames! There cant be any duplicate atomnames!) Move the atom name one space to the left in the PDB if it becomes 4 digits (eg HM10 in lysine should be moved one space toward \"ATOM\" rather than having the \"0\" go into the first resname column):\n",
    "                - N-term cap residue:\n",
    "                    - CA  -> CN1\n",
    "                    - C -> CN2\n",
    "                    - O -> ON1\n",
    "                    - H0# -> HN#\n",
    "                - Middle residue:\n",
    "                    - N -> NM1\n",
    "                    - H0# -> HM#\n",
    "                - C-term cap residue:\n",
    "                    - N -> NC1\n",
    "                    - CA -> CC1\n",
    "                    - H0# -> HC#\n",
    "            4. Rename ligand residue, change ligand chain to B, rename H's, and renumber atoms and residues (code below)\n",
    "            5. Load back into pymol and save as mol2.\n",
    "        - Ligand by itself (ligand coordinates)\n",
    "            1. From a duplicate of the ligand_res pdb, use pymol builder to remove prot atoms and change linked state back to unlinked state while affecting pose as little as possible. Note: Any H's affected by pymol builder will have to be renamed back to IUPAC standard.\n",
    "            2. Save as pdb. When saving, enable the following:\n",
    "                - Write multiple bonds as duplicate CONECT records\n",
    "                - Write CONECT records for all bonds\n",
    "                - Write segment identifier (segi) column\n",
    "            3. Rename added atoms (hydrogen for example) if needed and renumber atoms and residues if needed (code below).\n",
    "            4. Load back into pymol and save as mol2.\n",
    "            \n",
    "   - Check: \n",
    "            - Bond orders and connectivity in ligand mol2's are correct "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#function for renaming PDB H's \n",
    "#has an atomnumber limit of 99999 \n",
    "\n",
    "def renameH(pdbfile):\n",
    "\n",
    "    hatm_l=list()\n",
    "    tempfile=pdbfile+\".temp\"\n",
    "    tempfh = open(tempfile,\"w+\")\n",
    "    with open(pdbfile, 'r') as pdbfh :\n",
    "        lines = pdbfh.readlines()\n",
    "    print('Connecting Atom\\tOld Name\\tNew Name')\n",
    "    for hatmline in lines:\n",
    "        if hatmline.startswith('ATOM'):\n",
    "            oldname=hatmline[11:16].strip()\n",
    "            if oldname.startswith('H'):\n",
    "                hatmnbr=hatmline[9:11].strip()\n",
    "                for conline in lines:\n",
    "                    if conline.startswith('CONECT'):\n",
    "                        if hatmnbr == conline[6:11].strip():\n",
    "                            catmnbr = conline[11:16].strip()\n",
    "                            for catmline in lines:\n",
    "                                if catmline.startswith('ATOM'):\n",
    "                                    if catmnbr == catmline[9:11].strip():\n",
    "                                        catmname=catmline[11:16].strip()\n",
    "                                        catmname_suffix=catmname[1:]\n",
    "                                        n=str(1)\n",
    "                                        newname='H'+catmname_suffix+n\n",
    "                                        while newname in hatm_l:\n",
    "                                            n=str(int(n)+1)\n",
    "                                            newname='H'+catmname_suffix+n\n",
    "                                        hatm_l.append(newname)\n",
    "                                        hatmline = hatmline[:11]+newname.rjust(5)+hatmline[16:]\n",
    "                                        \n",
    "                                        print('{0}\\t\\t{1}\\t\\t{2}'.format(catmname,oldname, newname))\n",
    "        tempfh.write(hatmline)    \n",
    "    tempfh.close()\n",
    "    !mv $tempfile $pdbfile\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def alter_pdb(pdbfile,oldresname,newresname,oldchain):\n",
    "    print('\\nAltering {}'.format(pdbfile))\n",
    "    !cp $pdbfile \"$pdbfile\".bu\n",
    "    print(\"Made backup of \"+pdbfile)\n",
    "\n",
    "    #remove TER lines\n",
    "    fh = open(pdbfile, \"r\")\n",
    "    lines = fh.readlines()\n",
    "    fh.close()\n",
    "\n",
    "    fh = open(pdbfile, \"w\")\n",
    "    for line in lines:\n",
    "        if line[0:3] != \"TER\":\n",
    "            fh.write(line)\n",
    "    fh.close()\n",
    "    print(\"Removed TER lines\")\n",
    "    \n",
    "    #replace HETATM with ATOM\n",
    "     # Read in the file\n",
    "    with open(pdbfile, 'r') as pdbfh :\n",
    "        pdbfiledata = pdbfh.read()\n",
    "     # Replace the target string\n",
    "    pdbfiledata = pdbfiledata.replace('HETATM', 'ATOM  ')\n",
    "     # Write the file out again\n",
    "    with open(pdbfile, 'w') as pdbfh:\n",
    "        pdbfh.write(pdbfiledata)\n",
    "    print(\"Changed HETATM to ATOM\")\n",
    "    \n",
    "    #renumber atoms and residues \n",
    "    !pdb_reres $pdbfile | pdb_reatom > \"temp.pdb\"\n",
    "    !mv \"temp.pdb\" $pdbfile\n",
    "    print(\"Renumbered atoms and residues\")\n",
    "    \n",
    "    #change residue name and chain ID of ligand PDB\n",
    "    !pdb_rplresname -\"$oldresname\":\"$newresname\" $pdbfile | pdb_rplchain -\"$oldchain\":B > \"temp.pdb\"\n",
    "    !mv \"temp.pdb\" $pdbfile\n",
    "    print(\"Changed resname from {} to {} and changed chain from {} to B\".format(oldresname,newresname,oldchain))\n",
    "    \n",
    "    renameH(pdbfile)\n",
    "    print('Renamed hydrogens\\n')\n",
    "    return\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Altering IB_res.pdb\n",
      "Made backup of IB_res.pdb\n",
      "Removed TER lines\n",
      "Changed HETATM to ATOM\n",
      "Renumbered atoms and residues\n",
      "Changed resname from 8E8 to IB and changed chain from A to B\n",
      "Connecting Atom\tOld Name\tNew Name\n",
      "C2\t\tH21\t\tH21\n",
      "CAK\t\tHAK1\t\tHAK1\n",
      "CAM\t\tHAM1\t\tHAM1\n",
      "CAL\t\tHAL1\t\tHAL1\n",
      "CAN\t\tHAN1\t\tHAN1\n",
      "CAI\t\tHAI1\t\tHAI1\n",
      "CAF\t\tHAF1\t\tHAF1\n",
      "CAE\t\tHAE1\t\tHAE1\n",
      "CAG\t\tHAG1\t\tHAG1\n",
      "CAJ\t\tHAJ1\t\tHAJ1\n",
      "CAA\t\tHAA1\t\tHAA1\n",
      "CAA\t\tHAA2\t\tHAA2\n",
      "CAD\t\tHAD1\t\tHAD1\n",
      "CAD\t\tHAD2\t\tHAD2\n",
      "CAO\t\tHAO1\t\tHAO1\n",
      "CAO\t\tHAO2\t\tHAO2\n",
      "CAP\t\tHAP1\t\tHAP1\n",
      "CAP\t\tHAP2\t\tHAP2\n",
      "CAQ\t\tHAQ1\t\tHAQ1\n",
      "CAQ\t\tHAQ2\t\tHAQ2\n",
      "CAR\t\tHAR1\t\tHAR1\n",
      "CAR\t\tHAR2\t\tHAR2\n",
      "NAB\t\tHAB1\t\tHAB1\n",
      "NAB\t\tHAB2\t\tHAB2\n",
      "CBE\t\tHBE1\t\tHBE1\n",
      "CC3\t\tHC31\t\tHC31\n",
      "NN1\t\tHN11\t\tHN11\n",
      "CC4\t\tHC41\t\tHC41\n",
      "CB\t\tHB1\t\tHB1\n",
      "CB\t\tHB2\t\tHB2\n",
      "CC1\t\tHC11\t\tHC11\n",
      "CC3\t\tHC32\t\tHC32\n",
      "CC3\t\tHC33\t\tHC33\n",
      "NN2\t\tHN21\t\tHN21\n",
      "CC4\t\tHC42\t\tHC42\n",
      "CC4\t\tHC43\t\tHC43\n",
      "Renamed hydrogens\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#5.LINKED.D\n",
    "#manually cut and paste the ligand residue to the top of the PDB\n",
    "\n",
    "#MANUALLY CHANGE ARGUMENTS IN THIS COMMAND (FILE,OLDRESNAME,NEWRESNAME,OLDCHAIN)\n",
    "alter_pdb(\"IB_res.pdb\",\"8E8\",\"IB\",\"A\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Altering IB.pdb\n",
      "Made backup of IB.pdb\n",
      "Removed TER lines\n",
      "Changed HETATM to ATOM\n",
      "Renumbered atoms and residues\n",
      "Changed resname from 8E8 to IB and changed chain from B to B\n",
      "Connecting Atom\tOld Name\tNew Name\n",
      "CAA\t\tHAA1\t\tHAA1\n",
      "CAA\t\tHAA2\t\tHAA2\n",
      "CAD\t\tHAD1\t\tHAD1\n",
      "NAB\t\tHAB1\t\tHAB1\n",
      "NAB\t\tHAB2\t\tHAB2\n",
      "CAE\t\tHAE1\t\tHAE1\n",
      "CAF\t\tHAF1\t\tHAF1\n",
      "CAG\t\tHAG1\t\tHAG1\n",
      "CAI\t\tHAI1\t\tHAI1\n",
      "CAJ\t\tHAJ1\t\tHAJ1\n",
      "CAK\t\tHAK1\t\tHAK1\n",
      "CAL\t\tHAL1\t\tHAL1\n",
      "CAM\t\tHAM1\t\tHAM1\n",
      "CAN\t\tHAN1\t\tHAN1\n",
      "CAO\t\tHAO1\t\tHAO1\n",
      "CAO\t\tHAO2\t\tHAO2\n",
      "CAP\t\tHAP1\t\tHAP1\n",
      "CAP\t\tHAP2\t\tHAP2\n",
      "CAQ\t\tHAQ1\t\tHAQ1\n",
      "CAQ\t\tHAQ2\t\tHAQ2\n",
      "CAR\t\tHAR1\t\tHAR1\n",
      "CAR\t\tHAR2\t\tHAR2\n",
      "CBE\t\tHBE1\t\tHBE1\n",
      "C2\t\tH21\t\tH21\n",
      "Renamed hydrogens\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#5.UNLINKED.C\n",
    "#manually change desired atom name and cut and paste the line to the desired spot. \n",
    "    #eg. change added hydrogen from H26 (last one) to H00 (now first)\n",
    "    \n",
    "#MANUALLY CHANGE FILENAME AND RESIDUE NAMES IN THIS COMMAND \n",
    "alter_pdb(\"IB.pdb\",\"8E8\",\"IB\",\"B\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 6. Parameterize ligand"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  \n",
    "- Using CGenFF:\n",
    "    - Ligand by itself (ligand coordinates)\n",
    "        1. ...\n",
    "\n",
    "    - Ligand with attached residue with caps\n",
    "        1. ...\n",
    "\n",
    "    - Ligand with attached residue with caps (for complete parameterization w/ hybrid atom types)\n",
    "        1. Use same ligand_res mol2.\n",
    "        2. \"Include parameters that are already in CGenFF\"\n",
    "- Note: if removing lone pairs for halogens, comment out LP lines and add the LP charge to the halogen it was connected to."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Checks for setup:\n",
    "- Ligand is aligned to step3 protein pdb\n",
    "- Ligand is positioned such that adding the covalent bond can be resolved by minimization (doesn't get tangled)\n",
    "- Ligand is resid 1\n",
    "- Atomic differences between unlinked ligand and linked ligand make sense. \n",
    "    - eg. The coordinates of the ligand hydrogen removed during linking match the coordinates of that atom in the unlinked pdb, ie the expected hydrogen was removed and this may cause the remaining hydrogens to be renamed/renumbered.\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Set general variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#General variables\n",
    "\n",
    "#ligand name\n",
    "lig_name = \"IB\"\n",
    "\n",
    "#ligand PDB (positioned correctly to prot_solvated_f)\n",
    "lig_pdb = lig_name+\".pdb\" \n",
    "\n",
    "#ligand + res name\n",
    "lig_res_name = \"IB_res\"\n",
    "\n",
    "#ligand + res PDB\n",
    "lig_res_pdb = lig_res_name+\".pdb\"\n",
    "\n",
    "#ligand + res name (includes all cgenff parameters)\n",
    "lig_res_allparams = \"IB_res_allparams\"\n",
    "\n",
    "prot_name = \"MKK7\"\n",
    "\n",
    "#resid of covalent residue on protein\n",
    "cov_resid = \"101\" #106 if counting missing residues\n",
    "\n",
    "patch_name = \"LIGP\"\n",
    "\n",
    "#charmm-gui step3 output pdb\n",
    "prot_solvated_f = \"charmm-gui-5163983106/step3_pbcsetup.pdb\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "## Renumber waters and change segid of waters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Made file MKK7_solvated.pdb\n"
     ]
    }
   ],
   "source": [
    "#xtal waters must have segid WATA and be TIP3\n",
    "\n",
    "#make copy in build dir\n",
    "!cp \"$prot_solvated_f\" \"$prot_name\"_solvated.pdb\n",
    "print(\"Made file \"+prot_name+\"_solvated.pdb\")\n",
    "prot_solvated_f=prot_name+\"_solvated.pdb\"\n",
    "\n",
    "#change segid\n",
    "with open(prot_solvated_f, 'r') as pdbfh :\n",
    "    pdbfiledata = pdbfh.read()\n",
    " # Replace the target string\n",
    "pdbfiledata = pdbfiledata.replace('WATA', 'SOLV')\n",
    " # Write the file out again\n",
    "with open(prot_solvated_f, 'w') as pdbfh:\n",
    "    pdbfh.write(pdbfiledata)\n",
    "\n",
    "#renumber waters\n",
    "def reres_wat(pdbfile):\n",
    "    tempfile=pdbfile+\".temp\"\n",
    "    tempfh = open(tempfile,\"w+\")\n",
    "    with open(pdbfile, 'r') as pdbfh :\n",
    "        lines = pdbfh.readlines()\n",
    "    resnbr=0\n",
    "    prev_res=None\n",
    "    for line in lines:\n",
    "        if line.startswith('ATOM'):\n",
    "            if line[72:76] == 'SOLV':\n",
    "                if prev_res != line[22:27]:\n",
    "                    resnbr=resnbr+1\n",
    "                prev_res=line[22:27]\n",
    "                line = line[:22]+str(resnbr).rjust(4).ljust(5)+line[27:]\n",
    "\n",
    "\n",
    "\n",
    "        tempfh.write(line)    \n",
    "    tempfh.close()\n",
    "    !mv $tempfile $pdbfile\n",
    "    return\n",
    "\n",
    "reres_wat(prot_solvated_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "## Segment system "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "\n",
       "#### Paste the following code in VMD Tcl console to segment MKK7_solvated.pdb\n",
       "\n",
       "```tcl\n",
       "cd /mnt/c/Users/Kyle/Desktop/Roux/kininhibs/main/MKK7_IB/build/segments\n",
       "source segmenter.tcl\n",
       "```\n",
       "\n",
       "##### or run the following cell to run the vmd script from command line:\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get segments\n",
    "#CHANGE ION SELECTIONS IF USING ANYTHING OTHER THAN K and Cl\n",
    "os.makedirs(\"segments\", exist_ok=True)\n",
    "\n",
    "segmenter='''\n",
    "mol addfile ../PROT_SOLVATED_F\n",
    "\n",
    "set protein_residues [atomselect top \"protein\"]\n",
    "set water_residues [atomselect top \"water\"]\n",
    "set pot_residues [atomselect top \"resname POT\"]\n",
    "set cla_residues [atomselect top \"resname CLA\"]\n",
    "\n",
    "#separate every 10k waters if the residue number doesnt go to 5 digits\n",
    "#this gets around 10k res limit/duplicate segid's for solvent but isnt needed atm\n",
    "#mol addfile segments/solv.pdb\n",
    "#set waterresidues [atomselect top \"water\"]\n",
    "#set newsegnames [list]\n",
    "#set postfix 0\n",
    "#for { set i 0 } { $i < [llength $waterresidues] } { incr i } {\n",
    "#    if { $i % 30000 == 0 } {\n",
    "#        incr postfix\n",
    "#    }\n",
    "#    lappend newsegnames W$postfix\n",
    "#}\n",
    "#$waterresidues set segname $newsegnames\n",
    "\n",
    "\n",
    "\n",
    "#change chain names (ligand will be B)\n",
    "$protein_residues set chain A\n",
    "$water_residues set chain C\n",
    "$pot_residues set chain D\n",
    "$cla_residues set chain E\n",
    "\n",
    "$protein_residues writepdb proa.pdb\n",
    "$water_residues writepdb solv.pdb\n",
    "$pot_residues writepdb pot.pdb\n",
    "$cla_residues writepdb cla.pdb\n",
    "\n",
    "exit\n",
    "\n",
    "'''\n",
    "segmenter = re.sub(\"PROT_SOLVATED_F\", prot_solvated_f, segmenter)\n",
    "\n",
    "\n",
    "\n",
    "with open(\"./segments/segmenter.tcl\", \"w\") as segmenter_fh:\n",
    "    segmenter_fh.write(segmenter)\n",
    "    \n",
    "!cp $lig_pdb './segments/'$lig_name'.pdb' \n",
    "\n",
    "md('''\n",
    "#### Paste the following code in VMD Tcl console to segment {0}\n",
    "\n",
    "```tcl\n",
    "cd {1}/segments\n",
    "source segmenter.tcl\n",
    "```\n",
    "\n",
    "##### or run the following cell to run the vmd script from command line:\n",
    "'''.format(prot_solvated_f, os.getcwd()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) http://www.ks.uiuc.edu/Research/vmd/                         \n",
      "Info) Email questions and bug reports to vmd@ks.uiuc.edu           \n",
      "Info) Please include this reference in published work using VMD:   \n",
      "Info)    Humphrey, W., Dalke, A. and Schulten, K., `VMD - Visual   \n",
      "Info)    Molecular Dynamics', J. Molec. Graphics 1996, 14.1, 33-38.\n",
      "Info) -------------------------------------------------------------\n",
      "Info) Multithreading available, 12 CPUs.\n",
      "Info)   CPU features: SSE2 SSE4.1 AVX AVX2 FMA F16 HT \n",
      "Info) Free system memory: 7003MB (42%)\n",
      "Info) No CUDA accelerator devices available.\n",
      "Info) Dynamically loaded 3 plugins in directory:\n",
      "Info) /usr/local/lib/vmd/plugins/LINUXAMD64/molfile\n",
      "ERROR) Illegal rendering mode: GLSL\n",
      "after#0\n",
      "after#1\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl C:/Users/Kyle/Desktop/utilities/vmdStore\n",
      "The vmdStore package could not be loaded:\n",
      "\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl C:/Users/Kyle/Desktop/utilities/vmdStore C:/Users/Kyle/Desktop/utilities/vmdStore/plugins/molUP\n",
      "The molUP package could not be loaded:\n",
      "\n",
      "Info) Using plugin pdb for structure file ../MKK7_solvated.pdb\n",
      "Info) Using plugin pdb for coordinates from file ../MKK7_solvated.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 75764\n",
      "Info)    Bonds: 52085\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 24031\n",
      "Info)    Waters: 23589\n",
      "Info)    Segments: 3\n",
      "Info)    Fragments: 23729   Protein: 1   Nucleic: 0\n",
      "Info) Finished with coordinate file ../MKK7_solvated.pdb.\n",
      "0\n",
      "atomselect0\n",
      "atomselect1\n",
      "atomselect2\n",
      "atomselect3\n",
      "Info) Opened coordinate file proa.pdb for writing.\n",
      "Info) Finished with coordinate file proa.pdb.\n",
      "Info) Opened coordinate file solv.pdb for writing.\n",
      "Info) Finished with coordinate file solv.pdb.\n",
      "Info) Opened coordinate file pot.pdb for writing.\n",
      "Info) Finished with coordinate file pot.pdb.\n",
      "Info) Opened coordinate file cla.pdb for writing.\n",
      "Info) Finished with coordinate file cla.pdb.\n",
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) Exiting normally.\n",
      "vmd > "
     ]
    }
   ],
   "source": [
    "os.chdir('segments')\n",
    "!vmd -dispdev text -e segmenter.tcl\n",
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Checks:\n",
    "- PDBs are ok  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) http://www.ks.uiuc.edu/Research/vmd/                         \n",
      "Info) Email questions and bug reports to vmd@ks.uiuc.edu           \n",
      "Info) Please include this reference in published work using VMD:   \n",
      "Info)    Humphrey, W., Dalke, A. and Schulten, K., `VMD - Visual   \n",
      "Info)    Molecular Dynamics', J. Molec. Graphics 1996, 14.1, 33-38.\n",
      "Info) -------------------------------------------------------------\n",
      "Info) Multithreading available, 12 CPUs.\n",
      "Info)   CPU features: SSE2 SSE4.1 AVX AVX2 FMA F16 HT \n",
      "Info) Free system memory: 6715MB (41%)\n",
      "Info) No CUDA accelerator devices available.\n",
      "libGL error: No matching fbConfigs or visuals found\n",
      "libGL error: failed to load driver: swrast\n",
      "Info) OpenGL renderer: NVIDIA GeForce GTX 1070 Ti/PCIe/SSE2\n",
      "Info)   Features: STENCIL MSAA(4) MDE MTX NPOT PP PS \n",
      "Info)   GLSL rendering mode is NOT available.\n",
      "Info)   Textures: 2-D (32768x32768), 3-D (16384x16384x16384), Multitexture (4)\n",
      "Info) Dynamically loaded 3 plugins in directory:\n",
      "Info) /usr/local/lib/vmd/plugins/LINUXAMD64/molfile\n",
      "Warning) OpenGL Programmable Shading not available.\n",
      "after#0\n",
      "after#1\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/tk8.5 /usr/local/lib/vmd/scripts/tk8.5/ttk /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl C:/Users/Kyle/Desktop/utilities/vmdStore\n",
      "The vmdStore package could not be loaded:\n",
      "\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/tk8.5 /usr/local/lib/vmd/scripts/tk8.5/ttk /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl C:/Users/Kyle/Desktop/utilities/vmdStore C:/Users/Kyle/Desktop/utilities/vmdStore/plugins/molUP\n",
      "The molUP package could not be loaded:\n",
      "\n",
      "segments/*.pdb\n",
      "segments/cla.pdb segments/IB.pdb segments/pot.pdb segments/proa.pdb segments/solv.pdb\n",
      " \n",
      "Loading segments/cla.pdb\n",
      "Info) Using plugin pdb for structure file segments/cla.pdb\n",
      "Info) Using plugin pdb for coordinates from file segments/cla.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Finished with coordinate file segments/cla.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 53\n",
      "Info)    Bonds: 0\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 53\n",
      "Info)    Waters: 0\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 53   Protein: 0   Nucleic: 0\n",
      " \n",
      " \n",
      "Loading segments/IB.pdb\n",
      "Info) Using plugin pdb for structure file segments/IB.pdb\n",
      "Info) Using plugin pdb for coordinates from file segments/IB.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Eliminating bonds duplicated from existing structure...\n",
      "Info) Finished with coordinate file segments/IB.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 57\n",
      "Info)    Bonds: 61\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 1\n",
      "Info)    Waters: 0\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 1   Protein: 0   Nucleic: 0\n",
      " \n",
      " \n",
      "Loading segments/pot.pdb\n",
      "Info) Using plugin pdb for structure file segments/pot.pdb\n",
      "Info) Using plugin pdb for coordinates from file segments/pot.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Finished with coordinate file segments/pot.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 60\n",
      "Info)    Bonds: 0\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 60\n",
      "Info)    Waters: 0\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 60   Protein: 0   Nucleic: 0\n",
      " \n",
      " \n",
      "Loading segments/proa.pdb\n",
      "Info) Using plugin pdb for structure file segments/proa.pdb\n",
      "Info) Using plugin pdb for coordinates from file segments/proa.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Finished with coordinate file segments/proa.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 5137\n",
      "Info)    Bonds: 5194\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 318\n",
      "Info)    Waters: 0\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 1   Protein: 1   Nucleic: 0\n",
      " \n",
      " \n",
      "Loading segments/solv.pdb\n",
      "Info) Using plugin pdb for structure file segments/solv.pdb\n",
      "Info) Using plugin pdb for coordinates from file segments/solv.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Finished with coordinate file segments/solv.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 56466\n",
      "Info)    Bonds: 37644\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 18822\n",
      "Info)    Waters: 18822\n",
      "Info)    Segments: 1\n",
      "Info)    Fragments: 18822   Protein: 0   Nucleic: 0\n",
      " \n",
      "vmd > Info) In any publication of scientific results based in part or\n",
      "Info) completely on the use of the program STRIDE, please reference:\n",
      "Info)  Frishman,D & Argos,P. (1995) Knowledge-based secondary structure\n",
      "Info)  assignment. Proteins: structure, function and genetics, 23, 566-579.\n",
      "Warning) OpenGL Programmable Shading not available.\n",
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) Exiting normally.\n"
     ]
    }
   ],
   "source": [
    "# check segmented pdbs\n",
    "\n",
    "check='''\n",
    "set env(MYARG) \"segments/*.pdb\"\n",
    "\n",
    "set filelist [glob $env(MYARG)]\n",
    "\n",
    "foreach i $filelist {\n",
    "  echo \" \"\n",
    "  echo \"Loading $i\"\n",
    "  mol new \"$i\" type {pdb} waitfor all\n",
    "  echo \" \"\n",
    "}  \n",
    "'''\n",
    "\n",
    "with open(\"./segments/check.tcl\", \"w\") as check_fh:\n",
    "    check_fh.write(check)\n",
    "!vmd -e segments/check.tcl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "## Separate .rtf and .prm chunks from .str's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_str(f_nosuffix):\n",
    "    with open(f_nosuffix+\".str\",\"r\") as str_file,open(f_nosuffix+\".rtf\", 'w') as rtf_file,open(f_nosuffix+\".prm\", 'w') as prm_file:\n",
    "        rtfmode=False\n",
    "        prmmode=False\n",
    "        for line in str_file:\n",
    "            if \"read rtf card append\" in line:\n",
    "                rtfmode=True\n",
    "                prmmode=False\n",
    "            if \"read param card flex append\" in line:\n",
    "                rtfmode=False\n",
    "                prmmode=True\n",
    "            if rtfmode:\n",
    "                rtf_file.write(line)\n",
    "            elif prmmode:\n",
    "                prm_file.write(line)\n",
    "            else:\n",
    "                continue\n",
    "        rtf_file.write(\"\\n\"+\"RETURN\")\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#run the function here. use filename without the suffix\n",
    "split_str(lig_name)\n",
    "split_str(lig_res_name)\n",
    "split_str(lig_res_allparams)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find atoms with different atom types between two .rtf files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATOM <pdb name> <atom type in f1> <atom type in f2>\n",
      "ATOM CAA    CG321  CG2DC3\n",
      "ATOM CAD    CG321  CG2DC1\n",
      "ATOM HAA1   HGA2   HGA5  \n",
      "ATOM HAA2   HGA2   HGA5  \n",
      "ATOM HAD1   HGA2   HGA4  \n"
     ]
    }
   ],
   "source": [
    "f1=lig_res_name+\".rtf\" #can change variable\n",
    "f2=lig_name+\".rtf\" #can change variable\n",
    "\n",
    "\n",
    "fh1=open(f1,\"r\")\n",
    "fh2=open(f2, 'r')\n",
    "lines1 = fh1.readlines()\n",
    "lines2 = fh2.readlines()\n",
    "print(\"ATOM <pdb name> <atom type in f1> <atom type in f2>\")\n",
    "for line1 in lines1:\n",
    "    #print(line1[19:25])\n",
    "    for line2 in lines2:\n",
    "        if \"ATOM\" in line1 and \"ATOM\" in line2:\n",
    "            if line1[5:12].strip() == line2[5:12].strip():\n",
    "                if line1[12:18] != line2[12:18]:\n",
    "                    print(line1[0:18]+\" \"+line2[12:18])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Establish atom type dictionary and apply it to misc files if needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CG321 ': 'CT2   ', 'HGA2  ': 'HA2   ', 'CG311 ': 'CT1   '}\n"
     ]
    }
   ],
   "source": [
    "#dictionary for switching atom types\n",
    "#determined by which atoms will be found the protein side in the patch for dihedral terms \n",
    "\n",
    "\n",
    "cgenff_charmm_dict = {\"CG321\":\"CT2\", #CB\n",
    "                     \"HGA2\":\"HA2\",\n",
    "                     \"CG311\":\"CT1\"} #CA\n",
    "#                     \"NG311\":\"NZ   \",\n",
    "\n",
    "\n",
    "#adjust for spaces\n",
    "cgenff_charmm_dict = {key.ljust(6): value.ljust(6) for key, value in cgenff_charmm_dict.items()}\n",
    "\n",
    "print(cgenff_charmm_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#sub atom types in existing files \n",
    "def rpl_atomtypes(filename):     \n",
    "    # open the source file and read it        \n",
    "    fh = open(filename, 'r')\n",
    "    subject = fh.read()\n",
    "    fh.close()\n",
    "    #replace\n",
    "    for k,v in cgenff_charmm_dict.items():\n",
    "        subject = re.sub(k, v, subject)\n",
    "    # write the file\n",
    "    f_out = open(filename, 'w')\n",
    "    f_out.write(subject)\n",
    "    f_out.close()\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#not needed \n",
    "#rpl_atomtypes(lig_name+\".rtf\")\n",
    "#rpl_atomtypes(lig_res_name+\".rtf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check charges from an .rtf file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Atom \t\tCharge\n",
      "CAA     \t -0.068\n",
      "CAD     \t -0.106\n",
      "CAE     \t -0.115\n",
      "CAF     \t -0.115\n",
      "CAG     \t -0.115\n",
      "CAI     \t -0.115\n",
      "CAJ     \t -0.115\n",
      "CAK     \t -0.118\n",
      "CAL     \t -0.118\n",
      "CAM     \t -0.117\n",
      "CAN     \t -0.117\n",
      "CAO     \t -0.152\n",
      "CAP     \t -0.155\n",
      "CAQ     \t  0.147\n",
      "CAR     \t  0.065\n",
      "CAW     \t  0.470\n",
      "CAY     \t  0.216\n",
      "CAZ     \t  0.216\n",
      "NAB     \t -0.765\n",
      "NAU     \t -0.674\n",
      "OAC     \t -0.533\n",
      "OAV     \t -0.432\n",
      "CBA     \t  0.064\n",
      "CBB     \t  0.193\n",
      "CBE     \t  0.159\n",
      "NBF     \t  0.014\n",
      "NBG     \t -0.660\n",
      "N1      \t -0.741\n",
      "C2      \t  0.498\n",
      "N3      \t -0.794\n",
      "C4      \t  1.010\n",
      "C5      \t -0.388\n",
      "C6      \t  0.483\n",
      "HAA1    \t  0.090\n",
      "HAA2    \t  0.090\n",
      "HAB1    \t  0.378\n",
      "HAB2    \t  0.378\n",
      "HAD1    \t  0.090\n",
      "HAD2    \t  0.090\n",
      "HAE1    \t  0.115\n",
      "HAF1    \t  0.115\n",
      "HAG1    \t  0.115\n",
      "HAI1    \t  0.115\n",
      "HAJ1    \t  0.115\n",
      "HAK1    \t  0.115\n",
      "HAL1    \t  0.115\n",
      "HAM1    \t  0.115\n",
      "HAN1    \t  0.115\n",
      "HAO1    \t  0.090\n",
      "HAO2    \t  0.090\n",
      "HAP1    \t  0.090\n",
      "HAP2    \t  0.090\n",
      "HAQ1    \t  0.090\n",
      "HAQ2    \t  0.090\n",
      "HAR1    \t  0.090\n",
      "HAR2    \t  0.090\n",
      "HBE1    \t  0.090\n",
      "H21     \t  0.128\n",
      "CN1     \t -0.271\n",
      "CN2     \t  0.540\n",
      "ON1     \t -0.506\n",
      "HN11    \t  0.090\n",
      "HN12    \t  0.090\n",
      "HN13    \t  0.090\n",
      "CA      \t  0.072\n",
      "C       \t  0.474\n",
      "O       \t -0.519\n",
      "CB      \t -0.051\n",
      "SG      \t -0.246\n",
      "NM1     \t -0.454\n",
      "HA1     \t  0.090\n",
      "HB1     \t  0.090\n",
      "HB2     \t  0.090\n",
      "HM11    \t  0.296\n",
      "CC1     \t -0.112\n",
      "NC1     \t -0.477\n",
      "HC11    \t  0.090\n",
      "HC12    \t  0.090\n",
      "HC13    \t  0.090\n",
      "HC14    \t  0.333\n",
      "Total: -0.0\n"
     ]
    }
   ],
   "source": [
    "#this cell adds up the charges of a molecule from an rtf file\n",
    "#f=lig_name+\".rtf\" #user assign\n",
    "f=lig_res_name+\".rtf\" #user assign\n",
    "\n",
    "fh=open(f,\"r\")\n",
    "lines = fh.readlines()\n",
    "charge_list=[]\n",
    "print('Atom \\t\\tCharge')\n",
    "for line in lines:\n",
    "    if \"ATOM\" in line:\n",
    "        print(line[5:12],'\\t',line[19:25])\n",
    "        charge_list.append(float(line[19:25]))\n",
    "print(\"Total:\",round(sum(charge_list),3))\n",
    "         "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find atoms with different charges between two .rtf files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATOM <pdb name> <atom type in f1> <charge in f1> ! <charge in f2>\n",
      "ATOM CAA    CG321  -0.068 ! -0.421\n",
      "ATOM CAD    CG321  -0.106 ! -0.154\n",
      "ATOM CAW    CG2O1   0.470 !  0.495\n",
      "ATOM OAC    OG2D1  -0.533 ! -0.514\n",
      "ATOM NBG    NG2S0  -0.660 ! -0.624\n",
      "ATOM HAA1   HGA2    0.090 !  0.210\n",
      "ATOM HAA2   HGA2    0.090 !  0.210\n",
      "ATOM HAD1   HGA2    0.090 !  0.150\n",
      "total charge difference ([f1] - [f2])= 0.021\n"
     ]
    }
   ],
   "source": [
    "f1=lig_res_name+\".rtf\"\n",
    "f2=lig_name+\".rtf\"\n",
    "\n",
    "fh1=open(f1,'r')\n",
    "fh2=open(f2, 'r')\n",
    "lines1 = fh1.readlines()\n",
    "lines2 = fh2.readlines()\n",
    "chargediff=0\n",
    "print(\"ATOM <pdb name> <atom type in f1> <charge in f1> ! <charge in f2>\")\n",
    "for line1 in lines1:\n",
    "    #print(line1[19:25])\n",
    "    for line2 in lines2:\n",
    "        if \"ATOM\" in line1 and \"ATOM\" in line2:\n",
    "            if line1[5:12].strip() == line2[5:12].strip():\n",
    "                if line1[19:25] != line2[19:25]:\n",
    "                    print(line1[0:27]+\" \"+line2[19:25])\n",
    "                    chargediff=chargediff+(round(float(line1[19:25]),3)-round(float(line2[19:25]),3))\n",
    "print(\"total charge difference ([f1] - [f2])=\",round(chargediff,3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find different parameters for the same atomtypes between two .prm files\n",
    "#### - Note that there could be multiple dihedral lines for the same four atoms. This cell assumes that patches wont change multiplicity in the dihedral term "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BONDS\n",
      "ANGLES\n",
      "DIHEDRALS\n",
      "IMPROPERS\n"
     ]
    }
   ],
   "source": [
    "f1=lig_res_name+\".prm\"\n",
    "f2=lig_name+\".prm\"\n",
    "\n",
    "fh1=open(f1,'r')\n",
    "fh2=open(f2, 'r')\n",
    "lines1 = fh1.readlines()\n",
    "lines2 = fh2.readlines()\n",
    "seg1=None\n",
    "seg2=None\n",
    "\n",
    "for line1 in lines1:\n",
    "    if \"BONDS\" in line1:\n",
    "        seg1=\"BONDS\"\n",
    "        print(seg1)\n",
    "    elif \"ANGLES\" in line1:\n",
    "        seg1=\"ANGLES\"\n",
    "        print(seg1)\n",
    "    elif \"DIHEDRALS\" in line1:\n",
    "        seg1=\"DIHEDRALS\"\n",
    "        print(seg1)\n",
    "    elif \"IMPROPERS\" in line1:\n",
    "        seg1=\"IMPROPERS\"\n",
    "        print(seg1)\n",
    "    for line2 in lines2:\n",
    "        line1 = line1.split(\"!\")[0] #take out comment\n",
    "        line2 = line2.split(\"!\")[0] #take out comment\n",
    "        if \"BONDS\" in line1:\n",
    "            seg2=\"BONDS\"\n",
    "        elif \"ANGLES\" in line2:\n",
    "            seg2=\"ANGLES\"\n",
    "        elif \"DIHEDRALS\" in line2:\n",
    "            seg2=\"DIHEDRALS\"\n",
    "        elif \"IMPROPERS\" in line2:\n",
    "            seg2=\"IMPROPERS\"\n",
    "        if seg1==seg2==\"BONDS\":\n",
    "            if line1[0:13] == line2[0:13]:\n",
    "                if line1[13:] != line2[13:]:\n",
    "                    print('{0}:\\n{1}'.format(f1,line1))\n",
    "                    print('{0}:\\n{1}'.format(f2,line2))\n",
    "                    print('\\n')\n",
    "        elif seg1==seg2==\"ANGLES\":\n",
    "            if line1[0:20] == line2[0:20]:\n",
    "                if line1[20:] != line2[20:]:\n",
    "                    print('{0}:\\n{1}'.format(f1,line1))\n",
    "                    print('{0}:\\n{1}'.format(f2,line2))\n",
    "                    print('\\n')\n",
    "        elif seg1==seg2==\"DIHEDRALS\":\n",
    "            if line1[0:27] == line2[0:27] and line1[40:41] == line2[40:41]: \n",
    "                if line1[27:] != line2[27:]:\n",
    "                    print('{0}:\\n{1}'.format(f1,line1))\n",
    "                    print('{0}:\\n{1}'.format(f2,line2))\n",
    "                    print('\\n')\n",
    "        elif seg1==seg2==\"IMPROPERS\":\n",
    "            if line1[0:27] == line2[0:27]:\n",
    "                if line1[27:] != line2[27:]:\n",
    "                    print('{0}:\\n{1}'.format(f1,line1))\n",
    "                    print('{0}:\\n{1}'.format(f2,line2))\n",
    "                    print('\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine the amount of charge needed to get back to integer charge\n",
    "- Note that chargediff represents the extra charge added by patch ie the charge that should be subtracted from another atom in fudging\n",
    "- This is considering the charges before and after patching ie before and after writing the patch .rtf below. Therefore it may be easier to write the structure of the patch .rtf first, come back to this second, and finalize the patch .rtf third.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-0.0\n"
     ]
    }
   ],
   "source": [
    "#syntax: <atom name> = <charge after patch> - <charge before patch>\n",
    "\n",
    "chargediff=0\n",
    "\n",
    "#add differences in charge of affected protein atoms \n",
    "#Comparing to CYS from top_all36_prot.rtf\n",
    "#chargediff += (-0.454) - (-0.47) #n (nh1)\n",
    "#chargediff += (0.296) - (0.31) #hn (h)\n",
    "#chargediff += (0.072) - (0.07) #ca (ct1)\n",
    "#chargediff += (0.09) - (0.09) #ha (hb1)\n",
    "chargediff += (-0.051) - (-0.11) #cb (ct2)\n",
    "#chargediff += (0.09) - (0.09) #hb1 (ha2)\n",
    "#chargediff += (0.09) - (0.09) #hb2 (ha2)\n",
    "chargediff += (-0.246) - (-0.23) #sg (s)\n",
    "#chargediff += (0.474) - (0.51) #c (c)\n",
    "#chargediff += (-0.519) - (-0.51) #o (o)\n",
    "\n",
    "#add differences in charge of affected ligand atoms \n",
    "chargediff += (-0.068) - (-0.421) #caa\n",
    "chargediff += (-0.106) - (-0.154) #cad\n",
    "chargediff += (0.470) - (0.495) #caw\n",
    "chargediff += (-0.533) - (-0.514) #oac\n",
    "chargediff += (-0.660) - (-0.624) #nbg\n",
    "chargediff += (0.090) - (0.210) #haa1\n",
    "chargediff += (0.090) - (0.210) #haa2\n",
    "chargediff += (0.090) - (0.150) #had1\n",
    "chargediff += (-0.426) - (-0.432) #oav (fudging)\n",
    "\n",
    "#subtract charge of deleted protein atoms \n",
    "chargediff -= (0.16) #hg1\n",
    "\n",
    "#subtract charge of deleted ligand atoms \n",
    "#chargediff -= () #\n",
    "\n",
    "#add charge of added protein atoms \n",
    "#chargediff += () #\n",
    "\n",
    "#add charge of added ligand atoms \n",
    "chargediff += (0.090) #had2\n",
    "\n",
    "#add charge of added solvent atoms \n",
    "#chargediff += (1) #k\n",
    "\n",
    "#subtract charge of deleted solvent atoms \n",
    "#chargediff -= (-1) #cl\n",
    "\n",
    "print(round(chargediff,3))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write the patch .rtf\n",
    "- Add number prefixes to atom names (1 or 2 in the order given in the psfgen command and fix atom types to reflect how they are in the linked state)\n",
    "- Add fudged charges "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IC HAD2 CAD  CAA  SG   1.0894 109.7617  -96.2506 120.2539  1.6055\n"
     ]
    }
   ],
   "source": [
    "#calculate IC lines\n",
    "#IMPORTANT: when adding IC lines to replace coordinates of an existing atom, \n",
    "    #you need to add a DELETE ATOM line to the patch rtf for atom1 in the IC line\n",
    "    #AND \n",
    "    #you need to add a BOND line for the first two atoms in the IC line since the bond was deleted too \n",
    "\n",
    "pdbfile=lig_res_name+'.pdb'\n",
    "#atomnames should be without number prefix (put them in after) and be listed in order as they will come out on the IC line and be 4 char including spaces. Generate IC lines for added or spatially-altered atoms\n",
    "atom1='HAD2'\n",
    "atom2='CAD '\n",
    "atom3='CAA '\n",
    "atom4='SG '\n",
    "\n",
    "with open(pdbfile, 'r') as pdbfh :\n",
    "    lines = pdbfh.readlines()\n",
    "for line in lines:\n",
    "    if line.startswith('ATOM'):\n",
    "        if atom1.strip()==line[11:16].strip():\n",
    "            a1_x=float(line[30:38].strip())\n",
    "            a1_y=float(line[38:46].strip())\n",
    "            a1_z=float(line[46:54].strip())\n",
    "        elif atom2.strip()==line[11:16].strip():\n",
    "            a2_x=float(line[30:38].strip())\n",
    "            a2_y=float(line[38:46].strip())\n",
    "            a2_z=float(line[46:54].strip())\n",
    "        elif atom3.strip()==line[11:16].strip():\n",
    "            a3_x=float(line[30:38].strip())\n",
    "            a3_y=float(line[38:46].strip())\n",
    "            a3_z=float(line[46:54].strip())\n",
    "        elif atom4.strip()==line[11:16].strip():\n",
    "            a4_x=float(line[30:38].strip())\n",
    "            a4_y=float(line[38:46].strip())\n",
    "            a4_z=float(line[46:54].strip())\n",
    "\n",
    "def calculateDistance(x1,y1,z1,x2,y2,z2):\n",
    "    dist = math.sqrt((x2 - x1)**2 + (y2 - y1)**2 + (z2 - z1)**2)\n",
    "    return dist\n",
    "\n",
    "def calculateAngle(x1,y1,z1,x2,y2,z2,x3,y3,z3):\n",
    "    a = np.array([x1, y1, z1])\n",
    "    b = np.array([x2, y2, z2])\n",
    "    c = np.array([x3, y3, z3])\n",
    "\n",
    "    ba = a - b\n",
    "    bc = c - b\n",
    "\n",
    "    cosine_angle = np.dot(ba, bc) / (np.linalg.norm(ba) * np.linalg.norm(bc))\n",
    "    angle = np.arccos(cosine_angle)\n",
    "\n",
    "    return np.degrees(angle)\n",
    "\n",
    "def calculateDihedral(p):\n",
    "    #praxeolitic formula\n",
    "    p0 = p[0]\n",
    "    p1 = p[1]\n",
    "    p2 = p[2]\n",
    "    p3 = p[3]\n",
    "\n",
    "    b0 = -1.0*(p1 - p0)\n",
    "    b1 = p2 - p1\n",
    "    b2 = p3 - p2\n",
    "\n",
    "    # normalize b1 so that it does not influence magnitude of vector\n",
    "    # rejections that come next\n",
    "    b1 /= np.linalg.norm(b1)\n",
    "\n",
    "    # vector rejections\n",
    "    # v = projection of b0 onto plane perpendicular to b1\n",
    "    #   = b0 minus component that aligns with b1\n",
    "    # w = projection of b2 onto plane perpendicular to b1\n",
    "    #   = b2 minus component that aligns with b1\n",
    "    v = b0 - np.dot(b0, b1)*b1\n",
    "    w = b2 - np.dot(b2, b1)*b1\n",
    "\n",
    "    # angle between v and w in a plane is the torsion angle\n",
    "    # v and w may not be normalized but that's fine since tan is y/x\n",
    "    x = np.dot(v, w)\n",
    "    y = np.dot(np.cross(b1, v), w)\n",
    "    return np.degrees(np.arctan2(y, x))\n",
    "\n",
    "atom_ar = np.array([\n",
    "                [ a1_x, a1_y, a1_z ],\n",
    "                [ a2_x, a2_y, a2_z ],\n",
    "                [ a3_x, a3_y, a3_z ],\n",
    "                [ a4_x, a4_y, a4_z ]\n",
    "                ])\n",
    "\n",
    "IC1=round(calculateDistance(a1_x,a1_y,a1_z,a2_x,a2_y,a2_z),4)\n",
    "IC2=round(calculateAngle(a1_x,a1_y,a1_z,a2_x,a2_y,a2_z,a3_x,a3_y,a3_z),4)\n",
    "IC3=round(calculateDihedral(atom_ar),4)\n",
    "IC4=round(calculateAngle(a2_x,a2_y,a2_z,a3_x,a3_y,a3_z,a4_x,a4_y,a4_z),4)\n",
    "IC5=round(calculateDistance(a3_x,a3_y,a3_z,a4_x,a4_y,a4_z),4)\n",
    "print('IC {0} {1} {2} {3}'.format(atom1,atom2,atom3,atom4)\n",
    "      +'{: 8.4f}'.format(IC1)\n",
    "      +'{: 9.4f}'.format(IC2)\n",
    "      +'{: 10.4f}'.format(IC3)\n",
    "      +'{: 9.4f}'.format(IC4)\n",
    "      +'{: 8.4f}'.format(IC5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "header = '''read rtf card append\n",
    "* Topologies generated by\n",
    "* CHARMM General Force Field (CGenFF) program version 2.5\n",
    "*\n",
    "36 1\n",
    "\n",
    "! \"penalty\" is the highest penalty score of the associated parameters.\n",
    "! Penalties lower than 10 indicate the analogy is fair; penalties between 10\n",
    "! and 50 mean some basic validation is recommended; penalties higher than\n",
    "! 50 indicate poor analogy and mandate extensive validation/optimization.\n",
    "'''\n",
    "\n",
    "#use https://pages.jh.edu/rschlei1/Random_stuff/publications/charmmbook.pdf for IC usage\n",
    "#assign protein atoms charmm atom types except for the linking atom (eg sulfur in cysteine), which stays cgenff atomtype\n",
    "\n",
    "patch_rtf = '''\n",
    "PRES PATCHNAME\t\t0.00\n",
    "ATOM 1CB     CT2    -0.051 ! -0.11 from top_all36_prot.rtf RESI CYS\n",
    "ATOM 1SG     SG311  -0.246 ! -0.23 from top_all36_prot.rtf RESI CYS\n",
    "ATOM 2CAA    CG321  -0.068 ! -0.421, was CG2DC3\n",
    "ATOM 2CAD    CG321  -0.106 ! -0.154, was CG2DC1\n",
    "ATOM 2CAW    CG2O1   0.470 !  0.495\n",
    "ATOM 2OAC    OG2D1  -0.533 ! -0.514\n",
    "ATOM 2NBG    NG2S0  -0.660 ! -0.624\n",
    "ATOM 2HAA1   HGA2    0.090 !  0.210, was HGA5\n",
    "ATOM 2HAA2   HGA2    0.090 !  0.210, was HGA5\n",
    "ATOM 2HAD1   HGA2    0.090 !  0.150, was HGA4\n",
    "ATOM 2OAV    OG301  -0.426 ! -0.432, fudged \n",
    "ATOM 2HAD2   HGA2    0.090 ! added\n",
    "\n",
    "BOND 1SG   2CAA\n",
    "\n",
    "BOND 2HAD1 2CAD\n",
    "BOND 2HAD2 2CAD\n",
    "BOND 2HAA1 2CAA\n",
    "BOND 2HAA2 2CAA\n",
    "\n",
    "DELETE ATOM 1HG1 \n",
    "DELETE ATOM 2HAA1\n",
    "DELETE ATOM 2HAA2\n",
    "DELETE ATOM 2HAD1\n",
    "\n",
    "IC 2HAD1 2CAD 2CAA 1SG    1.0901 110.0353   23.4104 120.2539  1.6055\n",
    "IC 2HAD2 2CAD 2CAA 1SG    1.0894 109.7617  -96.2506 120.2539  1.6055\n",
    "IC 2HAA1 2CAA 2CAD 2CAW   1.0900 103.1338 -102.2136 108.4486  1.5025\n",
    "IC 2HAA2 2CAA 2CAD 2CAW   1.0908 106.7508   22.1977 108.4486  1.5025\n",
    "\n",
    "END\n",
    "RETURN'''\n",
    "\n",
    "patch_rtf = re.sub(\"PATCHNAME\", patch_name, patch_rtf)\n",
    "\n",
    "\n",
    "#for k,v in cgenff_charmm_dict.items():\n",
    "#    patch_rtf = re.sub(k, v, patch_rtf)\n",
    "\n",
    "with open(patch_name+\".rtf\",\"w\") as patch_rtf_file:\n",
    "    patch_rtf_file.write(header)\n",
    "    patch_rtf_file.write(patch_rtf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "## Write the patch .prm\n",
    "- Acceptable to have extra entries in patch prm file (they won't be used)\n",
    "- Take linked prm file with all dihedrals but those dihedrals have cgenff names so you need to replace names\n",
    "- Add new bonds, angles, and dihedrals with new names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prm file of lig+res\n",
    "lig_res_prm_file = lig_res_name+\".prm\" \n",
    "\n",
    "#prm file of lig+res that includes all cgenff terms \n",
    "lig_res_prm_allparams_file = lig_res_allparams+\".prm\" \n",
    "\n",
    "#new atoms or changed atomtypes from patch.rtf (cgenff names). Any line from <lig_res>.prm \n",
    "    #that has any of these will be included.\n",
    "atypelist1=[\"SG311\",\"CG321\",\"HGA2\"]\n",
    "\n",
    "#Atom(s) that would require a line with hybridized atomtypes (eg. SG311 in a cysteine linkage \n",
    "    #bc there should be charmm atomtypes on one side of the SG311 and cgenff atomtypes on the other)\n",
    "atypelist2=[\"SG311\"]\n",
    "\n",
    "header = '''read param card flex append\n",
    "* Parameters generated by analogy by\n",
    "* CHARMM General Force Field (CGenFF) program version 2.5\n",
    "*\n",
    "\n",
    "! Penalties lower than 10 indicate the analogy is fair; penalties between 10\n",
    "! and 50 mean some basic validation is recommended; penalties higher than\n",
    "! 50 indicate poor analogy and mandate extensive validation/optimization.\n",
    "\n",
    "'''\n",
    "\n",
    "#WILL NEED TO MANUALLY :\n",
    "    # - ADD THE PARAMS DECLARED AS MISSING BY CHAMBER AND MANUALLY CHANGE THE ATOMTYPES TO \n",
    "        #MATCH THE MISSING PARAM'D ATOMS\n",
    "    # - CHECK REPLACED PARAMS AND REMOVE IF NECESSARY \n",
    "\n",
    "#general rules for choosing b/w cgenff and prot params in the event of missing params:\n",
    "    #dihedral: if 3 atom types are protein (eg CT2 instead of CG321), use protein. Otherwise, cgenff. \n",
    "    #angles and impropers: if central atom is protein, use protein. Otherwise, cgenff\n",
    "    #bonds: cgenff\n",
    "    #if params dont specifically exist in prot, use cgenff\n",
    "\n",
    "\n",
    "#angle format ex\n",
    "#####  #####  #####   ###.##    ###.## \n",
    "#dihedral format ex \n",
    "#####  #####  #####  #####    ###.####  #   ###.####\n",
    "\n",
    "extra_bonds = '''\n",
    "SG311    CT2   198.000     1.8180 ! PROT fitted to C-S s   9/26/92 (FL), (patched s)-(patched cb) \n",
    "'''\n",
    "extra_angles = '''\n",
    "SG311  CT2    CT1      58.000  112.5000 ! (patched s)-(patched cb)-ca, from par_all36m_prot.prm (same in cgenff)\n",
    "CT2    SG311  CG321    34.00    95.00 ! (patched cb)-(patched s)-caa, from par_all36_cgenff.prm (same in cgenff)\n",
    "SG311  CT2    HA2      46.10    111.30 ! from par_all36m_prot.prm (same in cgenff)\n",
    "'''\n",
    "\n",
    "\n",
    "extra_dihedrals = '''\n",
    "C      CT1    CT2    SG311      0.2400  1   180.00  !fitted cys params with sg311 nomenclature from par_all36m_prot.prm\n",
    "C      CT1    CT2    SG311      0.7500  2   180.00\n",
    "C      CT1    CT2    SG311      1.3500  3   180.00\n",
    "NH1    CT1    CT2    SG311      0.3400  1     0.00\n",
    "NH1    CT1    CT2    SG311      0.5000  2   180.00\n",
    "NH1    CT1    CT2    SG311      1.4300  3     0.00\n",
    "HB1    CT1    CT2    SG311      0.2000  3     0.00 ! From X    CT1  CT2  X\n",
    "\n",
    "\n",
    "CT1    CT2    SG311  CG321      0.2400  1   180.00 ! IB_res , from CG321 CG321 SG311 CG321, PENALTY= 0.6\n",
    "CT1    CT2    SG311  CG321      0.3700  3     0.00 ! IB_res , from CG321 CG321 SG311 CG321, PENALTY= 0.6\n",
    "CG321  CG321  SG311  CT2        0.2400  1   180.00 ! PROT expt. MeEtS,      3/26/92 (FL)\n",
    "CG321  CG321  SG311  CT2        0.3700  3     0.00 ! PROT expt. MeEtS,      3/26/92 (FL)\n",
    "HGA2   CG321  SG311  CT2        0.2800  3     0.00 ! PROT DTN 8/24/90\n",
    "HA2    CT2    SG311  CG321      0.2800  3     0.00 ! PROT DTN 8/24/90\n",
    "'''\n",
    "    \n",
    "\n",
    "with open(lig_res_prm_file, 'r') as f, open(patch_name+\".prm\", 'w') as patch_prm_file:\n",
    "    patch_prm_file.write(header)\n",
    "    for line in f:\n",
    "        if \"BONDS\" in line:\n",
    "            patch_prm_file.write(\"\\n\"+line)\n",
    "            seg=\"BONDS\"\n",
    "        elif \"ANGLES\" in line:\n",
    "            patch_prm_file.write(extra_bonds)\n",
    "            patch_prm_file.write(\"\\n\"+line)\n",
    "            seg=\"ANGLES\"\n",
    "        elif \"DIHEDRALS\" in line:\n",
    "            patch_prm_file.write(extra_angles)\n",
    "            patch_prm_file.write(\"\\n\"+line)\n",
    "            seg=\"DIHEDRALS\"\n",
    "        elif \"IMPROPERS\" in line:\n",
    "            patch_prm_file.write(extra_dihedrals)\n",
    "            patch_prm_file.write(\"\\n\"+line)\n",
    "            seg=\"IMPROPERS\"\n",
    "        if any(word in line for word in [\"END\",\"RETURN\"]):\n",
    "                patch_prm_file.write(\"\\n\"+line)     \n",
    "        orig_line = line\n",
    "        line = line.split(\"!\")[0]+\"\\n\"\n",
    "        \n",
    "        #this will make duplicate lines between lig_prm_file and patch_prm_file, but parmed does not double energy\n",
    "        if any(word in line for word in atypelist1):\n",
    "            patch_prm_file.write(line)\n",
    "            \n",
    "print('Lines you may want to add or permutate manually with hybrid atomtypes:')\n",
    "with open(lig_res_prm_allparams_file, 'r') as f:\n",
    "    for line in f:\n",
    "        if \"BONDS\" in line:\n",
    "            seg=\"BONDS\"\n",
    "            print('\\nBONDS')\n",
    "        elif \"ANGLES\" in line:\n",
    "            seg=\"ANGLES\"\n",
    "            print('\\nANGLES')\n",
    "        elif \"DIHEDRALS\" in line:\n",
    "            seg=\"DIHEDRALS\"\n",
    "            print('\\nDIHEDRALS')\n",
    "        elif \"IMPROPERS\" in line:\n",
    "            seg=\"IMPROPERS\"\n",
    "            print('\\nIMPROPERS')\n",
    "        orig_line = line\n",
    "        line = line.split(\"!\")[0]+\"\\n\"\n",
    "        if any(word in line for word in atypelist2): \n",
    "            print(orig_line.strip(\"\\n\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "#### Checks:\n",
    "- Make sure ligand resname is the same for the ligand .rtf and .pdb file(s). \n",
    "    - .rtf resname probably needs to be changed from the cgenff default."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "## Write the extra .prm\n",
    "- Contains modified terms from previous scans\n",
    "- Contains dummy terms "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#may change if cgenff prm is updated. These use jul21 version\n",
    "header = '''read param card flex append\n",
    "* Parameters generated by analogy by\n",
    "* CHARMM General Force Field (CGenFF) program version 2.5\n",
    "*\n",
    "\n",
    "! Penalties lower than 10 indicate the analogy is fair; penalties between 10\n",
    "! and 50 mean some basic validation is recommended; penalties higher than\n",
    "! 50 indicate poor analogy and mandate extensive validation/optimization.\n",
    "\n",
    "'''\n",
    "\n",
    "#angle format ex\n",
    "#####  #####  #####   ###.##    ###.## \n",
    "#dihedral format ex \n",
    "#####  #####  #####  #####    ###.####  #   ###.####\n",
    "\n",
    "extra_bonds = '''\n",
    "SG3O2  FGP1    185.00     1.6200 ! O44 , from CG321 SG3O2, penalty= 360, was 1.79 corrected to QM opt, -trayder\n",
    "'''\n",
    "extra_angles = '''\n",
    "'''\n",
    "extra_dihedrals = '''\n",
    "CG321  CG311  NG2R51 NG2R50     0.7550  1     90.00 ! frag1\n",
    "CG321  CG311  NG2R51 NG2R50     0.5700  2    300.00 ! frag1\n",
    "CG321  CG311  NG2R51 NG2R50     0.0450  3     60.00 ! frag1\n",
    "CG321  CG311  NG2R51 NG2R50     0.5900  4      0.00 ! frag1\n",
    "\n",
    "! CG2DC3 CG2DC1 CG2O1  OG2D1      1.0000  1      0.00 ! frag6\n",
    "! CG2DC3 CG2DC1 CG2O1  OG2D1      0.8000  2    180.00 ! frag6\n",
    "! CG2DC3 CG2DC1 CG2O1  OG2D1      0.0300  3    180.00 ! frag6\n",
    "! CG2DC3 CG2DC1 CG2O1  OG2D1      0.1300  4      0.00 ! frag6\n",
    "\n",
    "CG321 CG321 CG2O1  OG2D1     -2.0500  1    180.00 ! frag8\n",
    "CG321 CG321 CG2O1  OG2D1     -0.9000  2      0.00 ! frag8\n",
    "CG321 CG321 CG2O1  OG2D1     -0.2500  3    180.00 ! frag8\n",
    "CG321 CG321 CG2O1  OG2D1     -0.1500  4    180.00 ! frag8\n",
    "\n",
    "CG321 CG1T1 CG1T2 HGPAM1        0.0     1      0.0  ! term alkyne\n",
    "NG2S1 CG321 CG1T1 CG1T2         0.0     1      0.0  ! term alkyne\n",
    "HGA2  CG321 CG1T1 CG1T2         0.0     1      0.0  ! term alkyne\n",
    "\n",
    "NG1T1  CG1N1  CG2DC1 CG2DC1       0.0     1      0.0  ! cyano\n",
    "NG1T1  CG1N1  CG2DC1 CG2O1        0.0     1      0.0  ! cyano\n",
    "NG1T1  CG1N1  CG311  CG311        0.0     1      0.0  ! cyano\n",
    "NG1T1  CG1N1  CG311  CG2O1        0.0     1      0.0  ! cyano\n",
    "NG1T1  CG1N1  CG311  HGA1         0.0     1      0.0  ! cyano\n",
    "\n",
    "NG2R62 CG2R64 NG301  CG321      1.9800  2   180.00 ! From NG2R62 CG2R64 NG301  CG331 in cgenff (TMC, yxu) to replace the flavin piperazine dihedral (NG2R62 CG2R64 NG301  CG321      0.4200  2     0.00) \n",
    "'''\n",
    "extra_impropers = '''\n",
    "'''\n",
    "extra_nbfix = '''\n",
    "'''\n",
    "\n",
    "with open(\"extra\"+'.prm', 'w') as extra_prm_fh:\n",
    "    extra_prm_fh.write(header)\n",
    "    extra_prm_fh.write('\\nBONDS\\n')\n",
    "    extra_prm_fh.write(extra_bonds)\n",
    "    extra_prm_fh.write('\\nANGLES\\n')\n",
    "    extra_prm_fh.write(extra_angles)\n",
    "    extra_prm_fh.write('\\nDIHEDRALS\\n')\n",
    "    extra_prm_fh.write(extra_dihedrals)\n",
    "    extra_prm_fh.write('\\nIMPROPERS\\n')\n",
    "    extra_prm_fh.write(extra_impropers)\n",
    "    extra_prm_fh.write('\\nNBFIX\\n')\n",
    "    extra_prm_fh.write(extra_nbfix)\n",
    "    extra_prm_fh.write('\\nEND\\n')\n",
    "    extra_prm_fh.write('\\nRETURN\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "## Construct .psf's with psfgen in VMD's Tcl console\n",
    "- Names might have changed between versions so the pdbalias lines might be screwed\n",
    "- Make sure toppar exists with the required files\n",
    "- Terminal patches might need to be altered (eg glycine N-term needs GLYP, not NTER)\n",
    "- Balance charge that may have been changed between linked/unlinked by removing solvent ions. \n",
    "    - delatom <segid> [resid] [atomname]\n",
    "    - eg. \"delatom CLA 1\" if a charge is lost in linked state (eg LYS is neutralized)\n",
    "- Guessed coords in unlinked may imply bad MODELLER output with overlapping side chains. Use pymol to sculpt that side chain to a better spot (delete any non-xtal interfereing waters from prot_solvated_f) and replace the coords in CHARMM-prot_solvated_f. Then change prot_solvated_f to the new pdb, and do the cells that use this pdb again such as segments and psf generation. Non-clashed coords for the interfered side chains can maybe be found in the CHARMM-GUI step 2 pdb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'os' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [3]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mos\u001b[49m\u001b[38;5;241m.\u001b[39mmakedirs(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtoppar\u001b[39m\u001b[38;5;124m\"\u001b[39m, exist_ok\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      2\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39msystem(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmv $patch_name\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m.rtf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m toppar/\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      3\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39msystem(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmv $patch_name\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m.prm\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m toppar/\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'os' is not defined"
     ]
    }
   ],
   "source": [
    "os.makedirs(\"toppar\", exist_ok=True)\n",
    "!mv $patch_name\".rtf\" toppar/\n",
    "!mv $patch_name\".prm\" toppar/\n",
    "!mv $lig_name\".rtf\" toppar/\n",
    "!mv $lig_name\".prm\" toppar/\n",
    "!mv \"extra.prm\" toppar/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp ../../resources/toppar_c36_jul21_essential/* toppar/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a) Unlinked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "\n",
       "#### Paste the following code in VMD Tcl console to generate the .psf \n",
       "#### or run the next cell\n",
       "\n",
       "```tcl\n",
       "cd /mnt/c/Users/Kyle/Desktop/Roux/kininhibs/main/MKK7_IB/build\n",
       "source MKK7_IB_u.pgn\n",
       "```\n",
       "\n",
       "\n",
       "#### After loading *only* the pdb and psf into VMD...\n",
       "   - This gets the net charge of the psf:\n",
       "```tcl\n",
       "set sel [atomselect top all]\n",
       "set charge [vecsum [$sel get charge]]\n",
       "```\n",
       "\n",
       "\n",
       "   - This gets the charge of specific selection (eg. protein):\n",
       "```tcl\n",
       "puts [vecsum [[atomselect top \"protein\"] get charge]]\n",
       "```\n",
       "\n",
       "\n",
       "   - This writes a crd file:\n",
       "```tcl\n",
       "[atomselect top all] writecrd MKK7_IB_u.crd\n",
       "```\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pgn_contents_u = '''\n",
    "package require psfgen\n",
    "resetpsf\n",
    "\n",
    "topology ./toppar/top_all36_prot.rtf\n",
    "#topology ./toppar/par_all36m_prot.prm\n",
    "topology ./toppar/top_all36_cgenff.rtf\n",
    "#topology ./toppar/par_all36_cgenff.prm\n",
    "\n",
    "topology ./toppar/toppar_water_ions.str\n",
    "\n",
    "topology ./toppar/LIGNAME.rtf\n",
    "topology ./toppar/LIGNAME.prm\n",
    "\n",
    "pdbalias residue HOH TIP3\n",
    "pdbalias atom TIP3 O OH2\n",
    "pdbalias residue HIS HSD\n",
    "\n",
    "segment PROA {\n",
    "    first NTER\n",
    "    last CTER\n",
    "    pdb ./segments/proa.pdb\n",
    "}\n",
    "coordpdb ./segments/proa.pdb PROA\n",
    "\n",
    "segment LIG {pdb ./segments/LIGNAME.pdb}\n",
    "coordpdb ./segments/LIGNAME.pdb LIG\n",
    "\n",
    "segment SOLV {\n",
    "    auto none\n",
    "    pdb ./segments/solv.pdb\n",
    "}\n",
    "coordpdb ./segments/solv.pdb SOLV\n",
    "\n",
    "segment POT {pdb ./segments/pot.pdb}\n",
    "coordpdb ./segments/pot.pdb POT\n",
    "\n",
    "segment CLA {pdb ./segments/cla.pdb}\n",
    "coordpdb ./segments/cla.pdb CLA\n",
    "\n",
    "regenerate angles dihedrals \n",
    "guesscoord\n",
    "\n",
    "writepdb PROTNAME_LIGNAME_u.pdb\n",
    "writepsf x-plor PROTNAME_LIGNAME_u.psf\n",
    "\n",
    "echo \" \"\n",
    "mol new PROTNAME_LIGNAME_u.pdb type {pdb} waitfor all\n",
    "mol addfile PROTNAME_LIGNAME_u.psf type {psf} waitfor all\n",
    "set sel [atomselect top all]\n",
    "set charge [vecsum [$sel get charge]]\n",
    "echo \"Net charge: $charge\"\n",
    "echo \" \"\n",
    "\n",
    "exit\n",
    "'''\n",
    "pgn_contents_u = re.sub(\"LIGNAME\", lig_name, pgn_contents_u)\n",
    "pgn_contents_u = re.sub(\"PROTNAME\", prot_name, pgn_contents_u)\n",
    "\n",
    "with open(prot_name+\"_\"+lig_name+\"_u.pgn\", \"w\") as pgn_fh:\n",
    "    pgn_fh.write(pgn_contents_u)\n",
    "\n",
    "md('''\n",
    "#### Paste the following code in VMD Tcl console to generate the .psf \n",
    "#### or run the next cell\n",
    "\n",
    "```tcl\n",
    "cd {0}\n",
    "source {1}.pgn\n",
    "```\n",
    "\n",
    "\n",
    "#### After loading *only* the pdb and psf into VMD...\n",
    "   - This gets the net charge of the psf:\n",
    "```tcl\n",
    "set sel [atomselect top all]\n",
    "set charge [vecsum [$sel get charge]]\n",
    "```\n",
    "\n",
    "\n",
    "   - This gets the charge of specific selection (eg. protein):\n",
    "```tcl\n",
    "puts [vecsum [[atomselect top \"protein\"] get charge]]\n",
    "```\n",
    "\n",
    "\n",
    "   - This writes a crd file:\n",
    "```tcl\n",
    "[atomselect top all] writecrd {1}.crd\n",
    "```\n",
    "\n",
    "'''.format(os.getcwd(),prot_name+\"_\"+lig_name+\"_u\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) http://www.ks.uiuc.edu/Research/vmd/                         \n",
      "Info) Email questions and bug reports to vmd@ks.uiuc.edu           \n",
      "Info) Please include this reference in published work using VMD:   \n",
      "Info)    Humphrey, W., Dalke, A. and Schulten, K., `VMD - Visual   \n",
      "Info)    Molecular Dynamics', J. Molec. Graphics 1996, 14.1, 33-38.\n",
      "Info) -------------------------------------------------------------\n",
      "Info) Multithreading available, 12 CPUs.\n",
      "Info)   CPU features: SSE2 SSE4.1 AVX AVX2 FMA F16 HT \n",
      "Info) Free system memory: 7007MB (42%)\n",
      "Info) No CUDA accelerator devices available.\n",
      "Info) Dynamically loaded 3 plugins in directory:\n",
      "Info) /usr/local/lib/vmd/plugins/LINUXAMD64/molfile\n",
      "ERROR) Illegal rendering mode: GLSL\n",
      "after#0\n",
      "after#1\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl C:/Users/Kyle/Desktop/utilities/vmdStore\n",
      "The vmdStore package could not be loaded:\n",
      "\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl C:/Users/Kyle/Desktop/utilities/vmdStore C:/Users/Kyle/Desktop/utilities/vmdStore/plugins/molUP\n",
      "The molUP package could not be loaded:\n",
      "\n",
      "2.0\n",
      "psfgen) clearing structure, preserving topology and aliases\n",
      "psfgen) reading topology file ./toppar/top_all36_prot.rtf\n",
      "\n",
      "psfgen) >>>>>>>>CHARMM36 All-Hydrogen Topology File for Proteins <<<<<<\n",
      "psfgen) >>>>> Includes phi, psi cross term map (CMAP) correction <<<<<<<\n",
      "psfgen) >>>>>>>>>>>>>>>>>>>>>>>>>> May 2011 <<<<<<<<<<<<<<<<<<<<<<<<<<<<\n",
      "psfgen)  All comments to the CHARMM web site: www.charmm.org\n",
      "psfgen)              parameter set discussion forum\n",
      "psfgen) \n",
      "psfgen) Created by CHARMM version 36 1\n",
      "psfgen) cross-term entries present in topology definitions\n",
      "psfgen) reading topology file ./toppar/top_all36_cgenff.rtf\n",
      "\n",
      "psfgen)   --------------------------------------------------------------------------  *\n",
      "psfgen)           CGenFF: Topology for the Charmm General Force Field v. 4.6          *\n",
      "psfgen)                     for Small Molecule Drug Design                            *\n",
      "psfgen)   --------------------------------------------------------------------------  *\n",
      "psfgen) \n",
      "psfgen) Created by CHARMM version 36 1\n",
      "psfgen) reading topology file ./toppar/toppar_water_ions.str\n",
      "\n",
      "psfgen)  Toplogy and parameter information for water and ions.\n",
      "psfgen) \n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 35: set nat ?NATC\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 36: set app\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE IF.  Line 40: if \"@NAT\" ne \"?NATC\" if @nat ne 0 set app append\n",
      "\n",
      "psfgen) reading topology from stream file\n",
      "psfgen)  Topology for water and ions\n",
      "psfgen) \n",
      "psfgen) Created by CHARMM version 31 1\n",
      "psfgen) skipping parameters in stream file\n",
      "psfgen)  Parameters for water and ions\n",
      "psfgen) \n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 319: set para\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE IF.  Line 320: if ?NUMNODE gt 1 set para node 0\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 322: set wrn ?WRNLEV\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE IF.  Line 324: if \"@WRN\" eq \"?WRNLEV\" set wrn 5\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 325: set bom ?bomlev\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE WRNLEV.  Line 326: WRNLEV -1 @PARA\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE BOMLEV.  Line 327: BOMLEV -1 @PARA\n",
      "\n",
      "psfgen) skipping parameters in stream file\n",
      "psfgen)  NBFix between carboxylate and sodium\n",
      "psfgen) \n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE BOMLEV.  Line 383: BOMLEV @bom @PARA\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE WRNLEV.  Line 384: WRNLEV @wrn @PARA\n",
      "\n",
      "psfgen) reading topology file ./toppar/IB.rtf\n",
      "\n",
      "psfgen) reading topology from stream file\n",
      "psfgen)  Topologies generated by\n",
      "psfgen)  CHARMM General Force Field (CGenFF) program version 2.5\n",
      "psfgen) \n",
      "psfgen) Created by CHARMM version 36 1\n",
      "psfgen) reading topology file ./toppar/IB.prm\n",
      "\n",
      "psfgen) skipping parameters in stream file\n",
      "psfgen)  Parameters generated by analogy by\n",
      "psfgen)  CHARMM General Force Field (CGenFF) program version 2.5\n",
      "psfgen) \n",
      "psfgen) aliasing residue HOH to TIP3\n",
      "psfgen) aliasing residue TIP3 atom O to OH2\n",
      "psfgen) aliasing residue HIS to HSD\n",
      "psfgen) building segment PROA\n",
      "psfgen) setting patch for first residue to NTER\n",
      "psfgen) setting patch for last residue to CTER\n",
      "psfgen) reading residues from pdb file ./segments/proa.pdb\n",
      "psfgen) extracted 303 residues from pdb file\n",
      "Info: generating structure...\n",
      "psfgen) Info: skipping improper N-C-CA-HN at beginning of segment.\n",
      "psfgen) Info: skipping cross-term C-N-CA-C-N-CA-C-N at beginning of segment.\n",
      "psfgen) Info: skipping conformation C-N-CA-C at beginning of segment.\n",
      "psfgen) Info: skipping conformation C-CA-N-HN at beginning of segment.\n",
      "psfgen) Info: skipping bond C-N at end of segment.\n",
      "psfgen) Info: skipping improper C-CA-N-O at end of segment.\n",
      "psfgen) Info: skipping cross-term C-N-CA-C-N-CA-C-N at end of segment.\n",
      "psfgen) Info: skipping conformation CA-C-N-CA at end of segment.\n",
      "psfgen) Info: skipping conformation N-CA-C-O at end of segment.\n",
      "psfgen) Info: skipping conformation N-CA-C-N at end of segment.\n",
      "\n",
      "\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/proa.pdb for segment PROA\n",
      "psfgen) building segment LIG\n",
      "psfgen) reading residues from pdb file ./segments/IB.pdb\n",
      "psfgen) extracted 1 residues from pdb file\n",
      "Info: generating structure...\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/IB.pdb for segment LIG\n",
      "psfgen) building segment SOLV\n",
      "psfgen) disabling angle autogeneration\n",
      "psfgen) disabling dihedral autogeneration\n",
      "psfgen) reading residues from pdb file ./segments/solv.pdb\n",
      "psfgen) extracted 23589 residues from pdb file\n",
      "Info: generating structure...\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/solv.pdb for segment SOLV\n",
      "psfgen) building segment POT\n",
      "psfgen) reading residues from pdb file ./segments/pot.pdb\n",
      "psfgen) extracted 66 residues from pdb file\n",
      "Info: generating structure...\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/pot.pdb for segment POT\n",
      "psfgen) building segment CLA\n",
      "psfgen) reading residues from pdb file ./segments/cla.pdb\n",
      "psfgen) extracted 73 residues from pdb file\n",
      "Info: generating structure...\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/cla.pdb for segment CLA\n",
      "psfgen) regenerating all angles\n",
      "psfgen) regenerating all dihedrals\n",
      "psfgen) Info: guessing coordinates for 0 atoms (0 non-hydrogen)\n",
      "psfgen) Info: writing pdb file MKK7_IB_u.pdb\n",
      "psfgen) Info: pdb file complete.\n",
      "psfgen) Info: writing psf file MKK7_IB_u.psf\n",
      "psfgen) total of 75821 atoms\n",
      "psfgen) total of 75735 bonds\n",
      "psfgen) total of 32600 angles\n",
      "psfgen) total of 13186 dihedrals\n",
      "psfgen) total of 816 impropers\n",
      "psfgen) total of 0 explicit exclusions\n",
      "psfgen) Structure requires EXTended PSF format\n",
      "psfgen) total of 301 cross-terms\n",
      "psfgen) Info: psf file complete.\n",
      " \n",
      "Info) Using plugin pdb for structure file MKK7_IB_u.pdb\n",
      "Info) Using plugin pdb for coordinates from file MKK7_IB_u.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Finished with coordinate file MKK7_IB_u.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 75821\n",
      "Info)    Bonds: 52221\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 24032\n",
      "Info)    Waters: 23582\n",
      "Info)    Segments: 5\n",
      "Info)    Fragments: 23723   Protein: 1   Nucleic: 0\n",
      "0\n",
      "psfplugin) Detected a Charmm PSF file\n",
      "psfplugin) Detected a Charmm PSF EXTEnded file\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 75821\n",
      "Info)    Bonds: 75735\n",
      "Info)    Angles: 32600  Dihedrals: 13186  Impropers: 816  Cross-terms: 301\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 24032\n",
      "Info)    Waters: 23589\n",
      "Info)    Segments: 5\n",
      "Info)    Fragments: 23730   Protein: 1   Nucleic: 0\n",
      "0\n",
      "atomselect0\n",
      "4.034489393234253e-6\n",
      "Net charge: 4.034489393234253e-6\n",
      " \n",
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) Exiting normally.\n",
      "vmd > "
     ]
    }
   ],
   "source": [
    "!vmd -dispdev text -e \"$prot_name\"_\"$lig_name\"_u.pgn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b) Linked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "\n",
       "#### Paste the following two lines in VMD Tcl console to generate the .psf\n",
       "#### or run the next cell\n",
       "\n",
       "```tcl\n",
       "cd /mnt/c/Users/Kyle/Desktop/Roux/kininhibs/main/MKK7_IB/build\n",
       "source MKK7_IB_l.pgn\n",
       "```\n",
       "\n",
       "\n",
       "#### After loading *only* the pdb and psf into VMD...\n",
       "   - This gets the net charge of the psf:\n",
       "```tcl\n",
       "set sel [atomselect top all]\n",
       "set charge [vecsum [$sel get charge]]\n",
       "```\n",
       "\n",
       "\n",
       "   - This gets the charge of specific selection (eg. protein):\n",
       "```tcl\n",
       "puts [vecsum [[atomselect top \"protein\"] get charge]]\n",
       "```\n",
       "\n",
       "\n",
       "   - This writes a crd file:\n",
       "```tcl\n",
       "[atomselect top all] writecrd MKK7_IB_l.crd\n",
       "```\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "pgn_contents_l = '''\n",
    "package require psfgen\n",
    "resetpsf\n",
    "\n",
    "topology ./toppar/top_all36_prot.rtf\n",
    "#topology ./toppar/par_all36m_prot.prm\n",
    "topology ./toppar/top_all36_cgenff.rtf\n",
    "#topology ./toppar/par_all36_cgenff.prm\n",
    "\n",
    "\n",
    "topology ./toppar/toppar_water_ions.str\n",
    "\n",
    "topology ./toppar/LIGNAME.rtf\n",
    "#topology ./toppar/LIGNAME.prm\n",
    "topology ./toppar/PATCHNAME.rtf\n",
    "topology ./toppar/PATCHNAME.prm\n",
    "\n",
    "pdbalias residue HOH TIP3\n",
    "pdbalias atom TIP3 O OH2\n",
    "pdbalias residue HIS HSD\n",
    "\n",
    "segment PROA {\n",
    "    first NTER\n",
    "    last CTER\n",
    "    pdb ./segments/proa.pdb\n",
    "}\n",
    "coordpdb ./segments/proa.pdb PROA\n",
    "\n",
    "segment LIG {pdb ./segments/LIGNAME.pdb}\n",
    "coordpdb ./segments/LIGNAME.pdb LIG\n",
    "\n",
    "segment SOLV {\n",
    "    auto none\n",
    "    pdb ./segments/solv.pdb\n",
    "}\n",
    "coordpdb ./segments/solv.pdb SOLV\n",
    "\n",
    "segment POT {pdb ./segments/pot.pdb}\n",
    "coordpdb ./segments/pot.pdb POT\n",
    "\n",
    "segment CLA {pdb ./segments/cla.pdb}\n",
    "coordpdb ./segments/cla.pdb CLA\n",
    "\n",
    "patch PATCHNAME PROA:PROTRESID LIG:1\n",
    "regenerate angles dihedrals \n",
    "guesscoord\n",
    "\n",
    "writepdb PROTNAME_LIGNAME_l.pdb\n",
    "writepsf x-plor PROTNAME_LIGNAME_l.psf\n",
    "\n",
    "echo \" \"\n",
    "mol new PROTNAME_LIGNAME_l.pdb type {pdb} waitfor all\n",
    "mol addfile PROTNAME_LIGNAME_l.psf type {psf} waitfor all\n",
    "set sel [atomselect top all]\n",
    "set charge [vecsum [$sel get charge]]\n",
    "echo \"Net charge: $charge\"\n",
    "echo \" \"\n",
    "\n",
    "exit\n",
    "'''\n",
    "pgn_contents_l = re.sub(\"LIGNAME\", lig_name, pgn_contents_l)\n",
    "pgn_contents_l = re.sub(\"PROTNAME\", prot_name, pgn_contents_l)\n",
    "pgn_contents_l = re.sub(\"PROTRESID\", cov_resid, pgn_contents_l)\n",
    "pgn_contents_l = re.sub(\"PATCHNAME\", patch_name, pgn_contents_l)\n",
    "\n",
    "with open(prot_name+\"_\"+lig_name+\"_l.pgn\", \"w\") as pgn_fh:\n",
    "    pgn_fh.write(pgn_contents_l)\n",
    "\n",
    "md('''\n",
    "#### Paste the following two lines in VMD Tcl console to generate the .psf\n",
    "#### or run the next cell\n",
    "\n",
    "```tcl\n",
    "cd {0}\n",
    "source {1}.pgn\n",
    "```\n",
    "\n",
    "\n",
    "#### After loading *only* the pdb and psf into VMD...\n",
    "   - This gets the net charge of the psf:\n",
    "```tcl\n",
    "set sel [atomselect top all]\n",
    "set charge [vecsum [$sel get charge]]\n",
    "```\n",
    "\n",
    "\n",
    "   - This gets the charge of specific selection (eg. protein):\n",
    "```tcl\n",
    "puts [vecsum [[atomselect top \"protein\"] get charge]]\n",
    "```\n",
    "\n",
    "\n",
    "   - This writes a crd file:\n",
    "```tcl\n",
    "[atomselect top all] writecrd {1}.crd\n",
    "```\n",
    "\n",
    "'''.format(os.getcwd(),prot_name+\"_\"+lig_name+\"_l\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) http://www.ks.uiuc.edu/Research/vmd/                         \n",
      "Info) Email questions and bug reports to vmd@ks.uiuc.edu           \n",
      "Info) Please include this reference in published work using VMD:   \n",
      "Info)    Humphrey, W., Dalke, A. and Schulten, K., `VMD - Visual   \n",
      "Info)    Molecular Dynamics', J. Molec. Graphics 1996, 14.1, 33-38.\n",
      "Info) -------------------------------------------------------------\n",
      "Info) Multithreading available, 12 CPUs.\n",
      "Info)   CPU features: SSE2 SSE4.1 AVX AVX2 FMA F16 HT \n",
      "Info) Free system memory: 7003MB (42%)\n",
      "Info) No CUDA accelerator devices available.\n",
      "Info) Dynamically loaded 3 plugins in directory:\n",
      "Info) /usr/local/lib/vmd/plugins/LINUXAMD64/molfile\n",
      "ERROR) Illegal rendering mode: GLSL\n",
      "after#0\n",
      "after#1\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl C:/Users/Kyle/Desktop/utilities/vmdStore\n",
      "The vmdStore package could not be loaded:\n",
      "\n",
      "/usr/local/lib/vmd/scripts/tcl8.5 /usr/local/lib/vmd/scripts /usr/local/lib/lib /Projects/johns/tcl/8.5.6/lib_LINUXAMD64/lib /usr/local/lib/vmd/scripts/vmd /usr/local/lib/vmd/plugins/LINUXAMD64/tcl /usr/local/lib/vmd/plugins/noarch/tcl C:/Users/Kyle/Desktop/utilities/vmdStore C:/Users/Kyle/Desktop/utilities/vmdStore/plugins/molUP\n",
      "The molUP package could not be loaded:\n",
      "\n",
      "2.0\n",
      "psfgen) clearing structure, preserving topology and aliases\n",
      "psfgen) reading topology file ./toppar/top_all36_prot.rtf\n",
      "\n",
      "psfgen) >>>>>>>>CHARMM36 All-Hydrogen Topology File for Proteins <<<<<<\n",
      "psfgen) >>>>> Includes phi, psi cross term map (CMAP) correction <<<<<<<\n",
      "psfgen) >>>>>>>>>>>>>>>>>>>>>>>>>> May 2011 <<<<<<<<<<<<<<<<<<<<<<<<<<<<\n",
      "psfgen)  All comments to the CHARMM web site: www.charmm.org\n",
      "psfgen)              parameter set discussion forum\n",
      "psfgen) \n",
      "psfgen) Created by CHARMM version 36 1\n",
      "psfgen) cross-term entries present in topology definitions\n",
      "psfgen) reading topology file ./toppar/top_all36_cgenff.rtf\n",
      "\n",
      "psfgen)   --------------------------------------------------------------------------  *\n",
      "psfgen)           CGenFF: Topology for the Charmm General Force Field v. 4.6          *\n",
      "psfgen)                     for Small Molecule Drug Design                            *\n",
      "psfgen)   --------------------------------------------------------------------------  *\n",
      "psfgen) \n",
      "psfgen) Created by CHARMM version 36 1\n",
      "psfgen) reading topology file ./toppar/toppar_water_ions.str\n",
      "\n",
      "psfgen)  Toplogy and parameter information for water and ions.\n",
      "psfgen) \n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 35: set nat ?NATC\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 36: set app\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE IF.  Line 40: if \"@NAT\" ne \"?NATC\" if @nat ne 0 set app append\n",
      "\n",
      "psfgen) reading topology from stream file\n",
      "psfgen)  Topology for water and ions\n",
      "psfgen) \n",
      "psfgen) Created by CHARMM version 31 1\n",
      "psfgen) skipping parameters in stream file\n",
      "psfgen)  Parameters for water and ions\n",
      "psfgen) \n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 319: set para\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE IF.  Line 320: if ?NUMNODE gt 1 set para node 0\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 322: set wrn ?WRNLEV\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE IF.  Line 324: if \"@WRN\" eq \"?WRNLEV\" set wrn 5\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE SET.  Line 325: set bom ?bomlev\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE WRNLEV.  Line 326: WRNLEV -1 @PARA\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE BOMLEV.  Line 327: BOMLEV -1 @PARA\n",
      "\n",
      "psfgen) skipping parameters in stream file\n",
      "psfgen)  NBFix between carboxylate and sodium\n",
      "psfgen) \n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE BOMLEV.  Line 383: BOMLEV @bom @PARA\n",
      "\n",
      "psfgen) ERROR!  FAILED TO RECOGNIZE WRNLEV.  Line 384: WRNLEV @wrn @PARA\n",
      "\n",
      "psfgen) reading topology file ./toppar/IB.rtf\n",
      "\n",
      "psfgen) reading topology from stream file\n",
      "psfgen)  Topologies generated by\n",
      "psfgen)  CHARMM General Force Field (CGenFF) program version 2.5\n",
      "psfgen) \n",
      "psfgen) Created by CHARMM version 36 1\n",
      "psfgen) reading topology file ./toppar/LIGP.rtf\n",
      "\n",
      "psfgen) reading topology from stream file\n",
      "psfgen)  Topologies generated by\n",
      "psfgen)  CHARMM General Force Field (CGenFF) program version 2.5\n",
      "psfgen) \n",
      "psfgen) Created by CHARMM version 36 1\n",
      "psfgen) reading topology file ./toppar/LIGP.prm\n",
      "\n",
      "psfgen) skipping parameters in stream file\n",
      "psfgen)  Parameters generated by analogy by\n",
      "psfgen)  CHARMM General Force Field (CGenFF) program version 2.5\n",
      "psfgen) \n",
      "psfgen) aliasing residue HOH to TIP3\n",
      "psfgen) aliasing residue TIP3 atom O to OH2\n",
      "psfgen) aliasing residue HIS to HSD\n",
      "psfgen) building segment PROA\n",
      "psfgen) setting patch for first residue to NTER\n",
      "psfgen) setting patch for last residue to CTER\n",
      "psfgen) reading residues from pdb file ./segments/proa.pdb\n",
      "psfgen) extracted 303 residues from pdb file\n",
      "Info: generating structure...\n",
      "psfgen) Info: skipping improper N-C-CA-HN at beginning of segment.\n",
      "psfgen) Info: skipping cross-term C-N-CA-C-N-CA-C-N at beginning of segment.\n",
      "psfgen) Info: skipping conformation C-N-CA-C at beginning of segment.\n",
      "psfgen) Info: skipping conformation C-CA-N-HN at beginning of segment.\n",
      "psfgen) Info: skipping bond C-N at end of segment.\n",
      "psfgen) Info: skipping improper C-CA-N-O at end of segment.\n",
      "psfgen) Info: skipping cross-term C-N-CA-C-N-CA-C-N at end of segment.\n",
      "psfgen) Info: skipping conformation CA-C-N-CA at end of segment.\n",
      "psfgen) Info: skipping conformation N-CA-C-O at end of segment.\n",
      "psfgen) Info: skipping conformation N-CA-C-N at end of segment.\n",
      "\n",
      "\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/proa.pdb for segment PROA\n",
      "psfgen) building segment LIG\n",
      "psfgen) reading residues from pdb file ./segments/IB.pdb\n",
      "psfgen) extracted 1 residues from pdb file\n",
      "Info: generating structure...\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/IB.pdb for segment LIG\n",
      "psfgen) building segment SOLV\n",
      "psfgen) disabling angle autogeneration\n",
      "psfgen) disabling dihedral autogeneration\n",
      "psfgen) reading residues from pdb file ./segments/solv.pdb\n",
      "psfgen) extracted 23589 residues from pdb file\n",
      "Info: generating structure...\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/solv.pdb for segment SOLV\n",
      "psfgen) building segment POT\n",
      "psfgen) reading residues from pdb file ./segments/pot.pdb\n",
      "psfgen) extracted 66 residues from pdb file\n",
      "Info: generating structure...\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/pot.pdb for segment POT\n",
      "psfgen) building segment CLA\n",
      "psfgen) reading residues from pdb file ./segments/cla.pdb\n",
      "psfgen) extracted 73 residues from pdb file\n",
      "Info: generating structure...\n",
      "segment complete.\n",
      "psfgen) reading coordinates from pdb file ./segments/cla.pdb for segment CLA\n",
      "psfgen) applying patch LIGP to 2 residue(s)\n",
      "\n",
      "psfgen) regenerating all angles\n",
      "psfgen) regenerating all dihedrals\n",
      "psfgen) Info: guessing coordinates for 4 atoms (0 non-hydrogen)\n",
      "psfgen) Info: writing pdb file MKK7_IB_l.pdb\n",
      "psfgen) Info: Atoms with guessed coordinates will have occupancy of 0.0.\n",
      "psfgen) Info: pdb file complete.\n",
      "psfgen) Info: writing psf file MKK7_IB_l.psf\n",
      "psfgen) total of 75821 atoms\n",
      "psfgen) total of 75736 bonds\n",
      "psfgen) total of 32606 angles\n",
      "psfgen) total of 13196 dihedrals\n",
      "psfgen) total of 816 impropers\n",
      "psfgen) total of 0 explicit exclusions\n",
      "psfgen) Structure requires EXTended PSF format\n",
      "psfgen) total of 301 cross-terms\n",
      "psfgen) Info: psf file complete.\n",
      " \n",
      "Info) Using plugin pdb for structure file MKK7_IB_u.pdb\n",
      "Info) Using plugin pdb for coordinates from file MKK7_IB_u.pdb\n",
      "Info) Determining bond structure from distance search ...\n",
      "Info) Finished with coordinate file MKK7_IB_u.pdb.\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 75821\n",
      "Info)    Bonds: 52221\n",
      "Info)    Angles: 0  Dihedrals: 0  Impropers: 0  Cross-terms: 0\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 24032\n",
      "Info)    Waters: 23582\n",
      "Info)    Segments: 5\n",
      "Info)    Fragments: 23723   Protein: 1   Nucleic: 0\n",
      "0\n",
      "psfplugin) Detected a Charmm PSF file\n",
      "psfplugin) Detected a Charmm PSF EXTEnded file\n",
      "Info) Analyzing structure ...\n",
      "Info)    Atoms: 75821\n",
      "Info)    Bonds: 75735\n",
      "Info)    Angles: 32600  Dihedrals: 13186  Impropers: 816  Cross-terms: 301\n",
      "Info)    Bondtypes: 0  Angletypes: 0  Dihedraltypes: 0  Impropertypes: 0\n",
      "Info)    Residues: 24032\n",
      "Info)    Waters: 23589\n",
      "Info)    Segments: 5\n",
      "Info)    Fragments: 23730   Protein: 1   Nucleic: 0\n",
      "0\n",
      "atomselect0\n",
      "4.034489393234253e-6\n",
      "Net charge: 4.034489393234253e-6\n",
      " \n",
      "Info) VMD for LINUXAMD64, version 1.9.4a57 (April 27, 2022)\n",
      "Info) Exiting normally.\n",
      "vmd > "
     ]
    }
   ],
   "source": [
    "!vmd -dispdev text -e \"$prot_name\"_\"$lig_name\"_l.pgn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "## Generate .parm7 and .rst7 files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#strip non-xtal water within <cutoff> angstroms of ligand\n",
    "#cutoff=5 is second water shell\n",
    "\n",
    "def calcDistance(x1,y1,z1,x2,y2,z2):\n",
    "    dist = math.sqrt((x2 - x1)**2 + (y2 - y1)**2 + (z2 - z1)**2)\n",
    "    return dist\n",
    "\n",
    "def stripwat(fn,prot_name,lig_name,cutoff):\n",
    "    \n",
    "    \n",
    "    #determine xtal wat residues\n",
    "    pdbfile=\"modeller/prot_wat.pdb\"\n",
    "\n",
    "    watres_l=['HOH','TIP3','WAT','TIP4']\n",
    "    with open(pdbfile, 'r') as pdbfh :\n",
    "        lines = pdbfh.readlines()\n",
    "    watcount=0\n",
    "    i=0\n",
    "    prev_res=9999999\n",
    "    for line in lines:\n",
    "        res=line[22:26].strip()  \n",
    "        if line[16:21].strip() in watres_l:\n",
    "\n",
    "            if watcount==0:\n",
    "                firstx=float(line[30:38].strip())\n",
    "                firsty=float(line[38:46].strip())  \n",
    "                firstz=float(line[46:54].strip())  \n",
    "            if res != prev_res:\n",
    "                watcount+=1\n",
    "            #if lines[i+1][16:21].strip() not in watres_l:\n",
    "            if 'O' in line[11:16].strip():\n",
    "                lastx=float(line[30:38].strip())\n",
    "                lasty=float(line[38:46].strip())  \n",
    "                lastz=float(line[46:54].strip())  \n",
    "        i+=1\n",
    "        prev_res=line[22:26].strip()  \n",
    "    dist1=np.round(calcDistance(firstx,firsty,firstz,lastx,lasty,lastz),1)\n",
    "    print('Found {0} waters in {1}'.format(watcount,pdbfile))\n",
    "    \n",
    "    #change filenames\n",
    "    !mv -nv \"$fn\".parm7 \"$fn\"_badwater.parm7\n",
    "    !mv -nv \"$fn\".rst7 \"$fn\"_badwater.rst7\n",
    "    reformat=!cpptraj -p \"$fn\"_badwater.parm7 -y \"$fn\".pdb -x \"$fn\"temp.pdb\n",
    "    !mv \"$fn\"temp.pdb \"$fn\".pdb\n",
    "    \n",
    "    prev_res=999999\n",
    "    maxwat=watcount\n",
    "    pdbfile=fn+'.pdb'\n",
    "    with open(pdbfile, 'r') as pdbfh :\n",
    "        lines = pdbfh.readlines()\n",
    "    for line in lines:\n",
    "        res=line[22:26].strip()\n",
    "        if line[16:21].strip() in watres_l:\n",
    "            if maxwat==watcount:\n",
    "                firstx=float(line[30:38].strip())\n",
    "                firsty=float(line[38:46].strip())  \n",
    "                firstz=float(line[46:54].strip())  \n",
    "                firstres=line[22:26].strip()  \n",
    "            if res != prev_res:\n",
    "                watcount-=1\n",
    "            if watcount==0 and 'O' in line[11:16].strip():\n",
    "                lastx=float(line[30:38].strip())\n",
    "                lasty=float(line[38:46].strip())  \n",
    "                lastz=float(line[46:54].strip())  \n",
    "                lastres=line[22:26].strip() \n",
    "                break\n",
    "        prev_res=line[22:26].strip()  \n",
    "    dist2=np.round(calcDistance(firstx,firsty,firstz,lastx,lasty,lastz),1)\n",
    "                \n",
    "               \n",
    "            \n",
    "    xtal_resrange=firstres+'-'+lastres\n",
    "    print('Xtal water resrange is '+xtal_resrange)\n",
    "\n",
    "    if dist1 != dist2: \n",
    "        print('ERROR: Distance between first and last waters in {0} ({1}) is different than'.format(pdbfile,dist1)) \n",
    "        print('\\tthat in modeller/prot_wat.pdb ({0}). Exiting...'.format(dist2))\n",
    "        sys.exit()\n",
    "    \n",
    "    \n",
    "    #strip with cpptraj \n",
    "    strip_parm7_contents='''\n",
    "    parm {0}_badwater.parm7\n",
    "    reference {0}_badwater.rst7 1\n",
    "    parmstrip (:{1}<:5)&:WAT&!(:{2})\n",
    "    parmwrite out {0}.parm7\n",
    "    go\n",
    "    '''.format(fn,lig_name,xtal_resrange) \n",
    "    with open(\"strip_parm7.traj\", \"w\") as cpptrajfh:\n",
    "        cpptrajfh.write(strip_parm7_contents)\n",
    "    trajlog_parm=!cpptraj -i strip_parm7.traj\n",
    "    \n",
    "    strip_rst7_contents='''\n",
    "    parm {0}_badwater.parm7\n",
    "    trajin {0}_badwater.rst7\n",
    "    reference {0}_badwater.rst7 1\n",
    "    autoimage\n",
    "    strip (:{1}<:{2})&:WAT&!(:{3})\n",
    "    trajout {0}.rst7\n",
    "    go\n",
    "    '''.format(fn,lig_name,cutoff,xtal_resrange) \n",
    "    with open(\"strip_rst7.traj\", \"w\") as cpptrajfh:\n",
    "        cpptrajfh.write(strip_rst7_contents)\n",
    "    trajlog_rst=!cpptraj -i strip_rst7.traj\n",
    "    \n",
    "    #get new pdb and rename badwater files\n",
    "    !mv -nv \"$fn\".pdb \"$fn\"_badwater.pdb\n",
    "    !mv -nv \"$fn\".psf \"$fn\"_badwater.psf\n",
    "    pdbtraj=!cpptraj -p \"$fn\".parm7 -y \"$fn\".rst7 -x \"$fn\".pdb\n",
    "    print(\"Removing badwaters from pdb and making \"+fn+\".pdb\")\n",
    "    \n",
    "    print(fn+'.parm7 and '+fn+'.rst7 are ready for minim, etc')\n",
    "    \n",
    "    return trajlog_parm,trajlog_rst"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### a) Unlinked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "\n",
       "#### Run in the command line or next cell:\n",
       "```\n",
       "parmed -i MKK7_IB_u.parmed\n",
       "```\n",
       "#### Or copy/paste in parmed:\n",
       "```python\n",
       "chamber -psf MKK7_IB_u.psf -crd MKK7_IB_u.pdb -param toppar/par_all36m_prot.prm -param toppar/par_all36_cgenff.prm -str toppar/toppar_water_ions.str -param toppar/IB.prm nocondense -box 93,93,93\n",
       "\n",
       "#HMassRepartition\n",
       "\n",
       "outparm MKK7_IB_u.parm7 MKK7_IB_u.rst7\n",
       "```\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#writes parmed script\n",
    "u_fn = prot_name+\"_\"+lig_name+\"_u\"\n",
    "u_parmed_contents = '''\n",
    "chamber -psf {0}.psf -crd {0}.pdb \\\n",
    "-param toppar/par_all36m_prot.prm \\\n",
    "-param toppar/par_all36_cgenff.prm \\\n",
    "-str toppar/toppar_water_ions.str \\\n",
    "-param toppar/{2}.prm \\\n",
    "-param toppar/extra.prm \\\n",
    "nocondense -box {1}\n",
    "\n",
    "#HMassRepartition\n",
    "\n",
    "outparm {0}.parm7 {0}.rst7\n",
    "\n",
    "#quit\n",
    "\n",
    "'''.format(u_fn,pbc,lig_name)\n",
    "\n",
    "with open(u_fn+\".parmed\", \"w\") as parmed_fh:\n",
    "    parmed_fh.write(u_parmed_contents)\n",
    "    \n",
    "md('''\n",
    "#### Run in the command line or next cell:\n",
    "```\n",
    "parmed -i {0}.parmed\n",
    "```\n",
    "#### Or copy/paste in parmed:\n",
    "```python\n",
    "chamber -psf {0}.psf -crd {0}.pdb -param toppar/par_all36m_prot.prm -param toppar/par_all36_cgenff.prm -str toppar/toppar_water_ions.str -param toppar/{2}.prm -param extra.prm nocondense -box {1}\n",
    "\n",
    "#HMassRepartition\n",
    "\n",
    "outparm {0}.parm7 {0}.rst7\n",
    "```\n",
    "'''.format(u_fn,pbc,lig_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      ".     .       .  .   . .   .   . .    +  .\n",
      "  .     .  :     .    .. :. .___---------___.\n",
      "       .  .   .    .  :.:. _\".^ .^ ^.  '.. :\"-_. .\n",
      "    .  :       .  .  .:../:            . .^  :.:\\.\n",
      "        .   . :: +. :.:/: .   .    .        . . .:\\\n",
      " .  :    .     . _ :::/:               .  ^ .  . .:\\\n",
      "  .. . .   . - : :.:./.                        .  .:\\\n",
      "  .      .     . :..|:                    .  .  ^. .:|\n",
      "    .       . : : ..||        .                . . !:|\n",
      "  .     . . . ::. ::\\(                           . :)/\n",
      " .   .     : . : .:.|. ######              .#######::|\n",
      "  :.. .  :-  : .:  ::|.#######           ..########:|\n",
      " .  .  .  ..  .  .. :\\ ########          :######## :/\n",
      "  .        .+ :: : -.:\\ ########       . ########.:/\n",
      "    .  .+   . . . . :.:\\. #######       #######..:/\n",
      "      :: . . . . ::.:..:.\\           .   .   ..:/\n",
      "   .   .   .  .. :  -::::.\\.       | |     . .:/\n",
      "      .  :  .  .  .-:.\":.::.\\             ..:/ .~~~~~~~~~~~.\n",
      " .      -.   . . . .: .:::.:.\\.           .:/,: Prepare to  \\\n",
      ".   .   .  :      : ....::_:..:\\   ___.  :/ .       be       |\n",
      "   .   .  .   .:. .. .  .: :.:.:\\       :/   \\  P A R M E D |\n",
      "     +   .   .   : . ::. :.:. .:.|\\  .:/|      \\~~~~~~~~~~~'\n",
      "     .         +   .  .  ...:: ..|  --.:|\n",
      ".      . . .   .  .  . ... :..:..\"(  ..)\"\n",
      " .   .       .      :  .   .: ::/  .  .::\\\n",
      "\n",
      "ParmEd: a Parameter file Editor\n",
      "\n",
      "\n",
      "['MKK7_IB_u.parmed']\n",
      "Reading actions from MKK7_IB_u.parmed\n",
      "\n",
      "Creating chamber topology file from PSF MKK7_IB_u.psf, PAR files [toppar/par_all36m_prot.prm, toppar/par_all36_cgenff.prm, toppar/IB.prm], and STR files [toppar/toppar_water_ions.str]. Coords from MKK7_IB_u.pdb. Using CMAP. Box info [93.0, 93.0, 93.0, 90.0, 90.0, 90.0]. GB Radius set mbondi.\n",
      "ParameterWarning: Replacing angle ('CG321', 'NG2S0', 'CG321'), <AngleType; k=45.000, theteq=121.000> with <AngleType; k=40.000, theteq=121.000>\n",
      "ParameterWarning: Replacing dihedral <DihedralType; phi_k=0.420, per=3, phase=0.000,  scee=1.000, scnb=1.000> with <DihedralType; phi_k=0.320, per=3, phase=0.000,  scee=1.000, scnb=1.000>\n",
      "Outputting Amber topology file MKK7_IB_u.parm7 and restart MKK7_IB_u.rst7\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "!rm \"$u_fn\".parm7\n",
    "!rm \"$u_fn\".rst7\n",
    "!parmed -i \"$u_fn\".parmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 182 waters in xtal/xtal_prot_wat.pdb\n",
      "renamed 'MKK7_IB_u.parm7' -> 'MKK7_IB_u_badwater.parm7'\n",
      "renamed 'MKK7_IB_u.rst7' -> 'MKK7_IB_u_badwater.rst7'\n",
      "\n",
      "CPPTRAJ: Trajectory Analysis. V5.1.0\n",
      "    ___  ___  ___  ___\n",
      "     | \\/ | \\/ | \\/ | \n",
      "    _|_/\\_|_/\\_|_/\\_|_\n",
      "\n",
      "| Date/time: 05/04/22 01:06:07\n",
      "| Available memory: 7.089 GB\n",
      "\n",
      "\tReading 'MKK7_IB_u_badwater.parm7' as Amber Topology\n",
      "\tCHAMBER topology:\n",
      "               >>>> CHARMM36 All-Hydrogen Parameter File for Proteins <<<<<<<<<<\n",
      "\tRadius Set: modified Bondi radii (mbondi)\n",
      "\tReading 'MKK7_IB_u.pdb' as PDB\n",
      "Warning: In PDB file MKK7_IB_u.pdb: 23589 name mismatches with parm MKK7_IB_u_badwater.parm7.\n",
      "Warning: Box information present in topology but not in trajectory.\n",
      "Warning: DISABLING BOX in topology 'MKK7_IB_u_badwater.parm7'!\n",
      "\tWriting 'MKK7_IB_utemp.pdb' as PDB\n",
      "---------- RUN BEGIN -------------------------------------------------\n",
      "\n",
      "PARAMETER FILES (1 total):\n",
      " 0: MKK7_IB_u_badwater.parm7, 75821 atoms, 24032 res, box: None, 23730 mol, 23589 solvent\n",
      "\n",
      "INPUT TRAJECTORIES (1 total):\n",
      " 0: 'MKK7_IB_u.pdb' is a PDB file, Parm MKK7_IB_u_badwater.parm7 (reading 1 of 1)\n",
      "  Coordinate processing will occur on 1 frames.\n",
      "\n",
      "OUTPUT TRAJECTORIES (1 total):\n",
      "  'MKK7_IB_utemp.pdb' (1 frames) is a PDB file\n",
      "\n",
      "BEGIN TRAJECTORY PROCESSING:\n",
      ".....................................................\n",
      "ACTIVE OUTPUT TRAJECTORIES (1):\n",
      "  MKK7_IB_utemp.pdb (coordinates)\n",
      "----- MKK7_IB_u.pdb (1-1, 1) -----\n",
      "100% Complete.\n",
      "\n",
      "Read 1 frames and processed 1 frames.\n",
      "TIME: Avg. throughput= 3.4172 frames / second.\n",
      "\n",
      "ACTION OUTPUT:\n",
      "TIME: Analyses took 0.0000 seconds.\n",
      "\n",
      "RUN TIMING:\n",
      "TIME:\t\tInit               : 0.0001 s (  0.02%)\n",
      "TIME:\t\tTrajectory Process : 0.2926 s ( 98.92%)\n",
      "TIME:\t\tAction Post        : 0.0000 s (  0.00%)\n",
      "TIME:\t\tAnalysis           : 0.0000 s (  0.00%)\n",
      "TIME:\t\tData File Write    : 0.0000 s (  0.00%)\n",
      "TIME:\t\tOther              : 0.0031 s (  0.01%)\n",
      "TIME:\tRun Total 0.2958 s\n",
      "---------- RUN END ---------------------------------------------------\n",
      "TIME: Total execution time: 0.4814 seconds.\n",
      "--------------------------------------------------------------------------------\n",
      "To cite CPPTRAJ use:\n",
      "Daniel R. Roe and Thomas E. Cheatham, III, \"PTRAJ and CPPTRAJ: Software for\n",
      "  Processing and Analysis of Molecular Dynamics Trajectory Data\". J. Chem.\n",
      "  Theory Comput., 2013, 9 (7), pp 3084-3095.\n",
      "\n",
      "Xtal water resrange is 305-486\n",
      "renamed 'MKK7_IB_u.pdb' -> 'MKK7_IB_u_badwater.pdb'\n",
      "renamed 'MKK7_IB_u.psf' -> 'MKK7_IB_u_badwater.psf'\n",
      "Removing badwaters from pdb and making MKK7_IB_u.pdb\n",
      "MKK7_IB_u.parm7 and MKK7_IB_u.rst7 are ready for minim, etc\n"
     ]
    }
   ],
   "source": [
    "#strip non-xtal water within <cutoff> angstroms of ligand\n",
    "#cutoff=5 is second water shell\n",
    "trajlog_parm,trajlog_rst=stripwat(u_fn,prot_name,lig_name,5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " 'CPPTRAJ: Trajectory Analysis. V5.1.0',\n",
       " '    ___  ___  ___  ___',\n",
       " '     | \\\\/ | \\\\/ | \\\\/ | ',\n",
       " '    _|_/\\\\_|_/\\\\_|_/\\\\_|_',\n",
       " '',\n",
       " '| Date/time: 05/04/22 01:06:08',\n",
       " '| Available memory: 7.072 GB',\n",
       " '',\n",
       " \"INPUT: Reading input from 'strip_parm7.traj'\",\n",
       " '  [parm MKK7_IB_u_badwater.parm7]',\n",
       " \"\\tReading 'MKK7_IB_u_badwater.parm7' as Amber Topology\",\n",
       " '\\tCHAMBER topology:',\n",
       " '               >>>> CHARMM36 All-Hydrogen Parameter File for Proteins <<<<<<<<<<',\n",
       " '\\tRadius Set: modified Bondi radii (mbondi)',\n",
       " '  [reference MKK7_IB_u_badwater.rst7 1]',\n",
       " \"\\tReading 'MKK7_IB_u_badwater.rst7' as Amber Restart\",\n",
       " \"\\tSetting active reference for distance-based masks: 'MKK7_IB_u_badwater.rst7'\",\n",
       " '  [parmstrip (:IB<:5)&:WAT&!(:305-486)]',\n",
       " '\\tStripping atoms in mask [(:IB<:5)&:WAT&!(:305-486)] (45) from MKK7_IB_u_badwater.parm7',\n",
       " '\\tStripped parm: 75776 atoms, 24017 res, box: Cubic, 23715 mol, 23574 solvent',\n",
       " '  [parmwrite out MKK7_IB_u.parm7]',\n",
       " \"\\tWriting topology 0 (MKK7_IB_u_badwater.parm7) to 'MKK7_IB_u.parm7' with format Amber Topology\",\n",
       " '\\tMemory used by full exclusion list: 390.680 kB',\n",
       " '\\tTopology has alternative residue numbering (from e.g PDB, stripping, reordering, etc).',\n",
       " '  [go]',\n",
       " '---------- RUN BEGIN -------------------------------------------------',\n",
       " 'Warning: No input trajectories specified.',\n",
       " 'TIME: Analyses took 0.0000 seconds.',\n",
       " '',\n",
       " 'RUN TIMING:',\n",
       " 'TIME:\\t\\tInit               : 0.0000 s (  0.00%)',\n",
       " 'TIME:\\t\\tTrajectory Process : 0.0000 s (  0.00%)',\n",
       " 'TIME:\\t\\tAction Post        : 0.0000 s (  0.00%)',\n",
       " 'TIME:\\t\\tAnalysis           : 0.0000 s (  0.00%)',\n",
       " 'TIME:\\t\\tData File Write    : 0.0000 s ( 10.00%)',\n",
       " 'TIME:\\t\\tOther              : 0.0000 s (  0.90%)',\n",
       " 'TIME:\\tRun Total 0.0000 s',\n",
       " '---------- RUN END ---------------------------------------------------',\n",
       " 'TIME: Total execution time: 0.4743 seconds.',\n",
       " '--------------------------------------------------------------------------------',\n",
       " 'To cite CPPTRAJ use:',\n",
       " 'Daniel R. Roe and Thomas E. Cheatham, III, \"PTRAJ and CPPTRAJ: Software for',\n",
       " '  Processing and Analysis of Molecular Dynamics Trajectory Data\". J. Chem.',\n",
       " '  Theory Comput., 2013, 9 (7), pp 3084-3095.',\n",
       " '']"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#uncomment log you want to check\n",
    "trajlog_parm\n",
    "#trajlog_rst"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b) Linked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "\n",
       "#### Run in the command line or next cell:\n",
       "```\n",
       "parmed -i MKK7_IB_l.parmed\n",
       "```\n",
       "#### Or copy/paste in parmed:\n",
       "```python\n",
       "chamber -psf MKK7_IB_l.psf -crd MKK7_IB_l.pdb -param toppar/par_all36m_prot.prm -param toppar/par_all36_cgenff.prm -str toppar/toppar_water_ions.str -param toppar/IB.prm -param toppar/LIGP.prm nocondense -box 93,93,93\n",
       "\n",
       "#HMassRepartition\n",
       "\n",
       "outparm MKK7_IB_l.parm7 MKK7_IB_l.rst7\n",
       "```\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#writes parmed script\n",
    "l_fn = prot_name+\"_\"+lig_name+\"_l\"\n",
    "l_parmed_contents = '''\n",
    "chamber -psf {0}.psf -crd {0}.pdb \\\n",
    "-param toppar/par_all36m_prot.prm \\\n",
    "-param toppar/par_all36_cgenff.prm \\\n",
    "-str toppar/toppar_water_ions.str \\\n",
    "-param toppar/{2}.prm \\\n",
    "-param toppar/{3}.prm \\\n",
    "-param toppar/extra.prm \\\n",
    "nocondense -box {1}\n",
    "\n",
    "#HMassRepartition\n",
    "\n",
    "outparm {0}.parm7 {0}.rst7\n",
    "\n",
    "#quit\n",
    "\n",
    "'''.format(l_fn,pbc,lig_name,patch_name)\n",
    "\n",
    "with open(l_fn+\".parmed\", \"w\") as parmed_fh:\n",
    "    parmed_fh.write(l_parmed_contents)\n",
    "    \n",
    "md('''\n",
    "#### Run in the command line or next cell:\n",
    "```\n",
    "parmed -i {0}.parmed\n",
    "```\n",
    "#### Or copy/paste in parmed:\n",
    "```python\n",
    "chamber -psf {0}.psf -crd {0}.pdb -param toppar/par_all36m_prot.prm -param toppar/par_all36_cgenff.prm -str toppar/toppar_water_ions.str -param toppar/{2}.prm -param toppar/{3}.prm -param extra.prm nocondense -box {1}\n",
    "\n",
    "#HMassRepartition\n",
    "\n",
    "outparm {0}.parm7 {0}.rst7\n",
    "```\n",
    "'''.format(l_fn,pbc,lig_name,patch_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                                  /^^^/           /]\n",
      "                                 /   ]           / ]\n",
      "                         _______/    ]___       /  ]\n",
      "                        /                \\_____/   /\n",
      "                      _/   [@]  \\ \\                \\\n",
      "                     /..         | |                ]\n",
      "                      VVVvvv\\    | |         _/\\    ]\n",
      "          P A R M E D       |               /    \\  ]\n",
      "                     AAA^^^^^              /       \\]\n",
      "                      \\_________\\   \\_____/\n",
      "                                \\    \\\n",
      "                                 \\____\\\n",
      "\n",
      "ParmEd: a Parameter file Editor\n",
      "\n",
      "\n",
      "['MKK7_IB_l.parmed']\n",
      "Reading actions from MKK7_IB_l.parmed\n",
      "\n",
      "Creating chamber topology file from PSF MKK7_IB_l.psf, PAR files [toppar/par_all36m_prot.prm, toppar/par_all36_cgenff.prm, toppar/IB.prm, toppar/LIGP.prm], and STR files [toppar/toppar_water_ions.str]. Coords from MKK7_IB_l.pdb. Using CMAP. Box info [93.0, 93.0, 93.0, 90.0, 90.0, 90.0]. GB Radius set mbondi.\n",
      "ParameterWarning: Replacing angle ('CG321', 'NG2S0', 'CG321'), <AngleType; k=45.000, theteq=121.000> with <AngleType; k=40.000, theteq=121.000>\n",
      "ParameterWarning: Replacing dihedral <DihedralType; phi_k=0.420, per=3, phase=0.000,  scee=1.000, scnb=1.000> with <DihedralType; phi_k=0.320, per=3, phase=0.000,  scee=1.000, scnb=1.000>\n",
      "Outputting Amber topology file MKK7_IB_l.parm7 and restart MKK7_IB_l.rst7\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "!rm \"$l_fn\".parm7\n",
    "!rm \"$l_fn\".rst7\n",
    "!parmed -i \"$l_fn\".parmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 182 waters in xtal/xtal_prot_wat.pdb\n",
      "renamed 'MKK7_IB_l.parm7' -> 'MKK7_IB_l_badwater.parm7'\n",
      "renamed 'MKK7_IB_l.rst7' -> 'MKK7_IB_l_badwater.rst7'\n",
      "\n",
      "CPPTRAJ: Trajectory Analysis. V5.1.0\n",
      "    ___  ___  ___  ___\n",
      "     | \\/ | \\/ | \\/ | \n",
      "    _|_/\\_|_/\\_|_/\\_|_\n",
      "\n",
      "| Date/time: 05/04/22 01:10:40\n",
      "| Available memory: 7.051 GB\n",
      "\n",
      "\tReading 'MKK7_IB_l_badwater.parm7' as Amber Topology\n",
      "\tCHAMBER topology:\n",
      "               >>>> CHARMM36 All-Hydrogen Parameter File for Proteins <<<<<<<<<<\n",
      "\tRadius Set: modified Bondi radii (mbondi)\n",
      "\tReading 'MKK7_IB_l.pdb' as PDB\n",
      "Warning: In PDB file MKK7_IB_l.pdb: 23589 name mismatches with parm MKK7_IB_l_badwater.parm7.\n",
      "Warning: Box information present in topology but not in trajectory.\n",
      "Warning: DISABLING BOX in topology 'MKK7_IB_l_badwater.parm7'!\n",
      "\tWriting 'MKK7_IB_ltemp.pdb' as PDB\n",
      "---------- RUN BEGIN -------------------------------------------------\n",
      "\n",
      "PARAMETER FILES (1 total):\n",
      " 0: MKK7_IB_l_badwater.parm7, 75821 atoms, 24032 res, box: None, 23729 mol, 23589 solvent\n",
      "\n",
      "INPUT TRAJECTORIES (1 total):\n",
      " 0: 'MKK7_IB_l.pdb' is a PDB file, Parm MKK7_IB_l_badwater.parm7 (reading 1 of 1)\n",
      "  Coordinate processing will occur on 1 frames.\n",
      "\n",
      "OUTPUT TRAJECTORIES (1 total):\n",
      "  'MKK7_IB_ltemp.pdb' (1 frames) is a PDB file\n",
      "\n",
      "BEGIN TRAJECTORY PROCESSING:\n",
      ".....................................................\n",
      "ACTIVE OUTPUT TRAJECTORIES (1):\n",
      "  MKK7_IB_ltemp.pdb (coordinates)\n",
      "----- MKK7_IB_l.pdb (1-1, 1) -----\n",
      "100% Complete.\n",
      "\n",
      "Read 1 frames and processed 1 frames.\n",
      "TIME: Avg. throughput= 3.2994 frames / second.\n",
      "\n",
      "ACTION OUTPUT:\n",
      "TIME: Analyses took 0.0000 seconds.\n",
      "\n",
      "RUN TIMING:\n",
      "TIME:\t\tInit               : 0.0000 s (  0.01%)\n",
      "TIME:\t\tTrajectory Process : 0.3031 s ( 98.95%)\n",
      "TIME:\t\tAction Post        : 0.0000 s (  0.00%)\n",
      "TIME:\t\tAnalysis           : 0.0000 s (  0.00%)\n",
      "TIME:\t\tData File Write    : 0.0000 s (  0.00%)\n",
      "TIME:\t\tOther              : 0.0032 s (  0.01%)\n",
      "TIME:\tRun Total 0.3063 s\n",
      "---------- RUN END ---------------------------------------------------\n",
      "TIME: Total execution time: 0.4962 seconds.\n",
      "--------------------------------------------------------------------------------\n",
      "To cite CPPTRAJ use:\n",
      "Daniel R. Roe and Thomas E. Cheatham, III, \"PTRAJ and CPPTRAJ: Software for\n",
      "  Processing and Analysis of Molecular Dynamics Trajectory Data\". J. Chem.\n",
      "  Theory Comput., 2013, 9 (7), pp 3084-3095.\n",
      "\n",
      "Xtal water resrange is 305-486\n",
      "renamed 'MKK7_IB_l.pdb' -> 'MKK7_IB_l_badwater.pdb'\n",
      "renamed 'MKK7_IB_l.psf' -> 'MKK7_IB_l_badwater.psf'\n",
      "Removing badwaters from pdb and making MKK7_IB_l.pdb\n",
      "MKK7_IB_l.parm7 and MKK7_IB_l.rst7 are ready for minim, etc\n"
     ]
    }
   ],
   "source": [
    "#strip non-xtal water within <cutoff> angstroms of ligand\n",
    "#cutoff=5 is second water shell\n",
    "trajlog_parm,trajlog_rst=stripwat(l_fn,prot_name,lig_name,5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " 'CPPTRAJ: Trajectory Analysis. V5.1.0',\n",
       " '    ___  ___  ___  ___',\n",
       " '     | \\\\/ | \\\\/ | \\\\/ | ',\n",
       " '    _|_/\\\\_|_/\\\\_|_/\\\\_|_',\n",
       " '',\n",
       " '| Date/time: 05/04/22 01:10:40',\n",
       " '| Available memory: 7.026 GB',\n",
       " '',\n",
       " \"INPUT: Reading input from 'strip_parm7.traj'\",\n",
       " '  [parm MKK7_IB_l_badwater.parm7]',\n",
       " \"\\tReading 'MKK7_IB_l_badwater.parm7' as Amber Topology\",\n",
       " '\\tCHAMBER topology:',\n",
       " '               >>>> CHARMM36 All-Hydrogen Parameter File for Proteins <<<<<<<<<<',\n",
       " '\\tRadius Set: modified Bondi radii (mbondi)',\n",
       " '  [reference MKK7_IB_l_badwater.rst7 1]',\n",
       " \"\\tReading 'MKK7_IB_l_badwater.rst7' as Amber Restart\",\n",
       " \"\\tSetting active reference for distance-based masks: 'MKK7_IB_l_badwater.rst7'\",\n",
       " '  [parmstrip (:IB<:5)&:WAT&!(:305-486)]',\n",
       " '\\tStripping atoms in mask [(:IB<:5)&:WAT&!(:305-486)] (45) from MKK7_IB_l_badwater.parm7',\n",
       " '\\tStripped parm: 75776 atoms, 24017 res, box: Cubic, 23714 mol, 23574 solvent',\n",
       " '  [parmwrite out MKK7_IB_l.parm7]',\n",
       " \"\\tWriting topology 0 (MKK7_IB_l_badwater.parm7) to 'MKK7_IB_l.parm7' with format Amber Topology\",\n",
       " '\\tMemory used by full exclusion list: 390.748 kB',\n",
       " '\\tTopology has alternative residue numbering (from e.g PDB, stripping, reordering, etc).',\n",
       " '  [go]',\n",
       " '---------- RUN BEGIN -------------------------------------------------',\n",
       " 'Warning: No input trajectories specified.',\n",
       " 'TIME: Analyses took 0.0000 seconds.',\n",
       " '',\n",
       " 'RUN TIMING:',\n",
       " 'TIME:\\t\\tInit               : 0.0000 s (  0.00%)',\n",
       " 'TIME:\\t\\tTrajectory Process : 0.0000 s (  0.00%)',\n",
       " 'TIME:\\t\\tAction Post        : 0.0000 s (  0.00%)',\n",
       " 'TIME:\\t\\tAnalysis           : 0.0000 s ( 12.50%)',\n",
       " 'TIME:\\t\\tData File Write    : 0.0000 s (  6.25%)',\n",
       " 'TIME:\\t\\tOther              : 0.0000 s (  0.81%)',\n",
       " 'TIME:\\tRun Total 0.0000 s',\n",
       " '---------- RUN END ---------------------------------------------------',\n",
       " 'TIME: Total execution time: 0.4877 seconds.',\n",
       " '--------------------------------------------------------------------------------',\n",
       " 'To cite CPPTRAJ use:',\n",
       " 'Daniel R. Roe and Thomas E. Cheatham, III, \"PTRAJ and CPPTRAJ: Software for',\n",
       " '  Processing and Analysis of Molecular Dynamics Trajectory Data\". J. Chem.',\n",
       " '  Theory Comput., 2013, 9 (7), pp 3084-3095.',\n",
       " '']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#uncomment log you want to check\n",
    "trajlog_parm\n",
    "#trajlog_rst"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check parm7, rst7, and pdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## QnA\n",
    "- Any general checks you like to do?\n",
    "    - Over first nanosecond, make sure bonds/dihedrals/etc are acting normal. eg doing eclipsed configs and are instead holding expected minima (gauche etc) \n",
    "    - QM investigation of dihedrals and bonds should be checked in theory (Gaussian). Evaluate energy of each rotation in a dihedral. Get stuff running first and come to this if there is any cause for concern.\n",
    "    - Check penalties. Could give wrong phase. If penalty is over cov bond, investigate more. Is it expected? Minima are in right locations? Good analogy if I have unsual atom environment?\n",
    "    - Check protonation states before charmmgui (protein) and paramchem (ligand). \n",
    "    - Check stereochemistry is right whenever you regenerate molecule through something (eg paramchem).\n",
    "    - Check that charge is integer\n",
    "    - Check that ring params are symmetrical and atomtypes are right\n",
    "    - Check ligand and protein clashes \n",
    "   \n",
    "- Is the charge-checking process right? Am I missing anything?\n",
    "    - Maintain symmetry when fudging. Dont make pos to neg and vice versa. \n",
    "    - Add ions if integer charge changes from patch. Can add with text editing. \n",
    "    \n",
    "- Will waters in the pocket be replaced by psfgen when combining the ligand and solvated protein? \n",
    "    - Waters that overlap can be fixed with visual inspection. They will usually be removed from the pocket durign equil. If system explodes, check. \n",
    "    \n",
    "- Is it fine to have the linked ligand be on a different chain? \n",
    "    - Probably. Chain isnt used in MD. \n",
    "\n",
    "- Do I fudge to get only ligand atoms to 0 total charge or do I include the difference in charge of patched side chain atoms?\n",
    "    - Result based.\n",
    "    - Can't go past 4 since cgenff doesnt go past 3 atoms. \n",
    "    - Avoid fudging anything on alpha carbon. \n",
    "    - Change beta carbon \n",
    "    \n",
    "- If aspartic acids are separated from solvent, local pH causes it to be protonated. Is this a pre-charmmgui/resname fix?\n",
    "    - Pre-charmmgui. Text edit PDB for protein. \n",
    "    \n",
    "- Psfgen doesnt put XPLOR in header even when writing as XPLOR format. Is this ok?\n",
    "    - Yes but errors will make it clear. \n",
    "    \n",
    "- Does my ligp.prm need anything added/removed?\n",
    "\n",
    "- When (not) to use nocondense in chamber?\n",
    "    - Removes duplicates dihedrals (check for details)\n",
    "    - "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TODO\n",
    "- decrease amount of dependencies \n",
    "- automate some setup steps as an option\n",
    "- improve \"set general variable\" section\n",
    "- add error detection and more print lines in general\n",
    "- does step3 pdb have to be tip3? (reres wat section)\n",
    "- automatically load checks into vmd \n",
    "- what does this mean in setup steps:\n",
    "    - #manually change desired atom name and cut and paste the line to the desired spot. \n",
    "    - #eg. change added hydrogen from H26 (last one) to H00 (now first)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
